From 42ca6111cd1ed332e58f317e3a6ac1d3824cc955 Mon Sep 17 00:00:00 2001
From: Shivani Raval <sraval@habana.ai>
Date: Tue, 01 Apr 2025 13:37:30 +0300
Subject: [PATCH] [JSW-31] build: added files for jgs build

As of now all the scripts are added same as FS1.
Asic build is passing and it is creating the
required binaries.

Change-Id: I6e652eb7fb8a1e261e5b45427b7032247c64d9ae
---

diff --git a/board/habanalabs/common/dirs.sh b/board/habanalabs/common/dirs.sh
index 9485deb..8f44b30 100644
--- a/board/habanalabs/common/dirs.sh
+++ b/board/habanalabs/common/dirs.sh
@@ -18,6 +18,9 @@
 fs1)
 	export SPECS_EMB_PROJ_DIR=${SPECS_EMBEDDED_ROOT}/fs/
 	;;
+jgs)
+	export SPECS_EMB_PROJ_DIR=${SPECS_EMBEDDED_ROOT}/fs/
+	;;
 *)
 	export SPECS_EMB_PROJ_DIR=${SPECS_EMBEDDED_ROOT}
 	;;
diff --git a/board/habanalabs/jgs/compute_d0_fake_fuse.txt b/board/habanalabs/jgs/compute_d0_fake_fuse.txt
new file mode 100644
index 0000000..b4b390d
--- /dev/null
+++ b/board/habanalabs/jgs/compute_d0_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/compute_d1_fake_fuse.txt b/board/habanalabs/jgs/compute_d1_fake_fuse.txt
new file mode 100644
index 0000000..b4b390d
--- /dev/null
+++ b/board/habanalabs/jgs/compute_d1_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/compute_d2_fake_fuse.txt b/board/habanalabs/jgs/compute_d2_fake_fuse.txt
new file mode 100644
index 0000000..b4b390d
--- /dev/null
+++ b/board/habanalabs/jgs/compute_d2_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/compute_d3_fake_fuse.txt b/board/habanalabs/jgs/compute_d3_fake_fuse.txt
new file mode 100644
index 0000000..b4b390d
--- /dev/null
+++ b/board/habanalabs/jgs/compute_d3_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/io_d0_fake_fuse.txt b/board/habanalabs/jgs/io_d0_fake_fuse.txt
new file mode 100644
index 0000000..56bb8f6
--- /dev/null
+++ b/board/habanalabs/jgs/io_d0_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00800000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/io_d1_fake_fuse.txt b/board/habanalabs/jgs/io_d1_fake_fuse.txt
new file mode 100644
index 0000000..b4b390d
--- /dev/null
+++ b/board/habanalabs/jgs/io_d1_fake_fuse.txt
@@ -0,0 +1,128 @@
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
+00000000
diff --git a/board/habanalabs/jgs/jgs-image-utility.sh b/board/habanalabs/jgs/jgs-image-utility.sh
new file mode 100755
index 0000000..ef02bb6
--- /dev/null
+++ b/board/habanalabs/jgs/jgs-image-utility.sh
@@ -0,0 +1,300 @@
+source ${BR2_EXTERNAL}/board/habanalabs/common/dirs.sh
+
+BINS_CONFIG_DIR="${BR2_EXTERNAL}/output/${HL_DEV}/images/bins_config"
+
+function swap_file {
+	local INF=$1
+	local OUTF=$2
+	local DBG=${3:-}
+
+	# SWAP needs this to be 4B aligned in size
+	dd status=none conv=sync bs=4 if=${INF} > ${OUTF}.tmp
+	# Sadly --endian=big is not swapping for -tx1 so I used some gawk magic
+	od -Anone -v -tx1 -w4 ${OUTF}.tmp | gawk '{ print $4 $3 $2 $1 }' | xxd -r -p > ${OUTF}
+
+	if [ -z "$DBG" ]; then
+		rm -f ${OUTF}.tmp
+	fi
+}
+
+function rev_bytes_stdin {
+	local str=$(cat -)
+
+	local len=${#str}
+	local reverse=""
+
+	[[ $len -eq 0 || $((len % 2)) -ne 0 ]] && return 1;
+	for ((i=((len - 2)); i>=0; i-=2)); do
+		reverse+=${str:$i:2}
+	done
+
+	echo -n $reverse
+}
+
+# get_image_property - returns value of a variable named: <name><property>
+function get_image_property {
+	local name=${1^^}
+	local property=${2^^}
+	local value=$(eval "echo \${${name}${property}}")
+
+	if [ -z "$value" ]; then
+		echo "Error: no \"${name}${property}\" value exist..."
+		exit 1;
+	else
+		echo $value
+	fi
+}
+
+function get_image_fname {
+	get_image_property $1 "_BINFILE"
+}
+
+function get_image_base {
+	get_image_property $1 "_OFFSET"
+}
+
+function get_image_loadaddr {
+	get_image_property $1 "_TEXT_BASE"
+}
+
+function get_image_startaddr {
+	get_image_property $1 "_START_ADDR"
+}
+
+function get_image_size {
+	local name=${1^^}
+
+	stat -Lc "%s" $(eval "echo \${${name}_BINFILE}")
+}
+
+function hl_zephyr_objcopy_sections() {
+	local PREFIX=$1
+	local MEM=$2
+	local ALIGN=$3
+	local SECTIONS=${*:4}
+
+	${OBJCOPY} -S --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON \
+	--remove-section=.eh_frame $(echo -n $SECTIONS | sed 's/[^ ]* */ --only-section=&/g') ${ELFS_DIR}/zephyr_${PREFIX}.elf \
+	--set-section-flags=bss=alloc,load,contents --set-section-flags=noinit=alloc,load,contents \
+	${STAGE_DIR}/zephyr-${PREFIX}-${MEM}.bin
+
+	dd status=none conv=sync bs=${ALIGN} if=${STAGE_DIR}/zephyr-${PREFIX}-${MEM}.bin > \
+	${BINS_DIR}/${DEV_PREFIX}-zephyr-${PREFIX}-${MEM}.bin
+}
+
+function hl_print_elf_sections_csv {
+	local app=$1
+
+	echo "name,size,address"
+	arc-linux-objdump -h ${ELFS_DIR}/zephyr_${app}.elf | \
+		awk '/^[ \t]+[0-9]+ [^.]/ { printf("%s,0x%s,0x%s\n", $2, $3, $4);}'
+}
+
+function hl_remove_empty_lines {
+	sed '/^[[:space:]]*$/d'
+}
+
+function hl_extract_zephyr_bins_sections {
+	local app=$1
+	local bins_file=$2
+	local sections=()
+	local bin_start
+	local bin_end
+	local sec_start
+	local sec_end
+
+	echo "Extracting sections for ${app} app"
+	while IFS=',' read -r bin_name bin_addr bin_size bin_align; do
+		echo -e "\tBinary ${bin_name} ($((${bin_size} * 1024))@${bin_addr})"
+		sections=()
+		bin_start=$((${bin_addr}))
+		bin_end=$((${bin_addr} + $((${bin_size} * 1024))))
+
+		while IFS=',' read -r sec_name sec_size sec_addr; do
+			sec_start=$((${sec_addr}))
+			sec_end=$((${sec_addr} + ${sec_size}))
+
+			if [ ${sec_start} -lt ${bin_start} -o ${sec_end} -gt ${bin_end} ]; then
+				continue
+			fi
+
+			sections+=(${sec_name})
+
+			echo -e "\t\tSection ${sec_name} ($((${sec_size}))@${sec_addr})";
+		done < <(hl_print_elf_sections_csv ${app} | tail -n +2)
+
+		hl_zephyr_objcopy_sections ${app} ${bin_name} ${bin_align} ${sections[@]}
+
+		echo ""
+
+	done < <(tail -n +2 ${bins_file} | hl_remove_empty_lines)
+}
+
+# Find a bins file which fits the app
+function hl_zephyr_get_bins {
+	local app=$1
+	local bins_file
+	local bins_app
+
+	# Try to find to closest match between an elf and its bins file
+	while [ $(echo ${app} | tr "_" " " | wc -w) -gt 0 ]; do
+		for bins_file in $(ls $BINS_CONFIG_DIR/${HL_TARGET}_*.bins); do
+			bins_app=$(echo ${bins_file} | sed "s/.*\/${HL_TARGET}_\(.*\)\.bins/\1/")
+
+			if [ "${bins_app}" == "${app}" ]; then
+				echo -n ${bins_file}
+				return
+			fi
+		done
+
+		# Remove last word
+		app=$(echo ${app} | awk -F_ 'BEGIN {OFS="_"} {NF-=1} 1')
+	done
+}
+
+# This function should be called after hl_create_zephyr_bins_config
+function hl_extract_zephyr_elf_sections {
+	local elf_file
+	local bins_file
+	local app
+
+	for elf_file in $(ls ${ELFS_DIR}/zephyr_*.elf); do
+		app=$(echo ${elf_file} | sed "s/.*\/zephyr_\(.*\)\.elf/\1/")
+		bins_file=$(hl_zephyr_get_bins ${app})
+
+		if [ -n "${bins_file}" ]; then
+			hl_extract_zephyr_bins_sections ${app} ${bins_file}
+			continue
+		fi
+	done
+}
+
+function hl_create_zephyr_bins_config {
+mkdir -p $BINS_CONFIG_DIR
+	rm $BINS_CONFIG_DIR/* -fr
+	mem_sections_hdr_path=${SPECS_EMB_PROJ_DIR}/${DEV_PREFIX}/${DEV_PREFIX}_mem_addr_dt.h
+
+	while read -r module; do
+		if echo "$module" | grep -q -E "iccm" || echo "$module" | grep -q -E "dccm"; then
+			module=$(echo "$module" | cut -d '_' -f 1)
+		fi
+
+		# TEMPORARY:
+		# Skip all the modules except "mgmt", because they are not implemented yet.
+		# Only a basic implementation of arcmgmt is being implemented for now.
+		if ! echo "$module" | grep -q -E  -e "mgmt" -e "agent"; then
+			continue;
+		fi
+
+		bins_path=$BINS_CONFIG_DIR/${HL_TARGET}_${module}.bins
+		echo "name,address,size_kb,align_bytes" > $bins_path
+		if echo "$module" | grep -q -E "io" ||  echo "$module" | grep -q -E "compute"; then
+			while read -r section; do
+				off=`grep ${module^^}_$section.*_OFFSET $mem_sections_hdr_path | sed 's/[[:space:]]\+/ /g' | cut -d ' ' -f3`
+				sz=`grep ${module^^}_$section.*_SIZE $mem_sections_hdr_path | sed 's/[[:space:]]\+/ /g' | cut -d ' ' -f3`
+				echo ${section,,},0x$off,$sz,16 >> $bins_path
+			done <<< `grep "_SECTION" $mem_sections_hdr_path | grep ${module^^} | cut -d '_' -f4| sort -u` #section
+		else
+			while read -r section; do
+				if [ "$section" == "ICCM" ] || [ "$section" == "DCCM" ]; then
+					off=`grep ${module^^}_$section.*_OFFSET $mem_sections_hdr_path | sed 's/[[:space:]]\+/ /g' | cut -d ' ' -f3`
+					sz=`grep ${module^^}_$section.*_SIZE $mem_sections_hdr_path | sed 's/[[:space:]]\+/ /g' | cut -d ' ' -f3`
+					echo ${section,,},0x$off,$sz,16 >> $bins_path
+				fi
+			done <<< `grep "_SECTION" $mem_sections_hdr_path | grep ${module^^} | cut -d '_' -f3`
+		fi
+	done <<< `grep "_SECTION" $mem_sections_hdr_path | cut -d '_' -f2,3 | tr [:upper:] [:lower:]`   #module}
+}
+
+function hl_create_section_variables {
+	local mem_sections_hdr_path=${SPECS_EMB_PROJ_DIR}/${DEV_PREFIX}/${DEV_PREFIX}_mem_addr_dt.h
+
+	while read line; do
+		export $line
+	done <<< `grep "define" $mem_sections_hdr_path | sed 's/[[:space:]]\+/ /g' | cut -d ' ' -f2-3 | sed  's/ /=/g'`
+}
+
+function get_struct_size {
+	header_file="$1"
+	struct_name="$2"
+
+	# Check if a filename is provided
+	if [ -z "$header_file" ]; then
+		echo "Usage: get_struct_size <header_file> <struct_name>"
+		exit 1
+	fi
+
+	# Check if a struct name is provided
+	if [ -z "$struct_name" ]; then
+		echo "Usage: get_struct_size <header_file> <struct_name>"
+		exit 1
+	fi
+
+	# Use gcc to preprocess the header file and calculate the size of the struct
+	size=$(echo "#include \"$header_file\"
+	#include <stdio.h>
+	int main() {
+		printf(\"%zu\", sizeof(struct $struct_name));
+		return 0;
+	}" | gcc -x c - -o tmp - && ./tmp)
+
+	if [ -z "$size" ]; then
+		echo "Error: Struct '$struct_name' not found in '$header_file'"
+		exit 1
+	else
+		echo $size
+	fi
+
+	# Clean up
+	rm -f tmp
+}
+
+function get_fake_fuse_bin_list_array {
+	HEADER_FILE="$1"
+	TARGET_ENUM="$2"
+
+	# Check if the header file exists
+	if [[ ! -f "$HEADER_FILE" ]]; then
+		echo "Header file '$HEADER_FILE' does not exist."
+		exit 1
+	fi
+
+	# Extract enum members for the specific enum using awk
+	ENUM_MEMBERS=$(awk -v enum_name="$TARGET_ENUM" '
+		BEGIN { inside_enum = 0 }
+		/enum[[:space:]]+[^[:space:]]*[[:space:]]*{/{
+			if ($0 ~ "enum[[:space:]]+" enum_name "[[:space:]]*{") {
+				inside_enum = 1
+			}
+		}
+		/}/ { if (inside_enum) { inside_enum = 0; next } }
+		inside_enum && /^[[:space:]]*[[:upper:]][[:alnum:]_]*/ {
+			gsub(/[[:space:]]*,?[[:space:]]*$/, "", $0)
+			print $0
+		}
+	' "$HEADER_FILE")
+
+	IFS=' ' read -r -a FAKE_FUSE_ARRAY <<< "$(echo "$ENUM_MEMBERS" | awk '{gsub(/^[[:space:]]+|[[:space:]]+$/, "", $0); gsub(/[[:space:]]+/, " ", $0); printf "%s ", $0}' | sed 's/ $//')"
+
+	export FAKE_FUSE_ARRAY
+}
+
+function hl_palad_hex {
+	local NAME=$(echo "$1" | cut -f 1 -d '.')
+	local FULLFILENAME=$2
+	local FILE=$(basename ${FULLFILENAME})
+	local BYTES_PER_LINE=$3
+	local DESTDIR=${4:-${HEXS_DIR}}
+	# pad file to 128bit (16Byte) as pladium emulates 128bit memory, so xxd wont trunc lines
+	dd status=none conv=sync bs=16 if=${FULLFILENAME} > ${STAGE_DIR}/${FILE}.tmp
+	xxd -e -g ${BYTES_PER_LINE} ${STAGE_DIR}/${FILE}.tmp | xxd -r | xxd -ps -c${BYTES_PER_LINE} > ${DESTDIR}/${DEV_PREFIX}-${HL_TARGET}-${NAME}.hex
+	rm ${STAGE_DIR}/${FILE}.tmp
+}
+
+function hl_palad_flash_hex {
+	local NAME=$(echo "$1" | cut -f 1 -d '.')
+	local FILE=$2
+	local DESTDIR=${3:-${HEXS_DIR}}
+	# save hex as a 1 byte list of values
+	xxd -e -g 4 ${FILE} | xxd -r | xxd -ps -c1 > ${DESTDIR}/${DEV_PREFIX}-${HL_TARGET}-${NAME}.hex
+}
diff --git a/board/habanalabs/jgs/jgs-local.mk b/board/habanalabs/jgs/jgs-local.mk
new file mode 100644
index 0000000..836f54a
--- /dev/null
+++ b/board/habanalabs/jgs/jgs-local.mk
@@ -0,0 +1,16 @@
+ARCBOOT_OVERRIDE_SRCDIR = $(TOPDIR)/../arcboot
+BAREMETAL_OVERRIDE_SRCDIR = $(TOPDIR)/../preboot
+ZEPHYR_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr
+#PID_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/arcpid
+MGMT_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/arcmgmt
+#MGMT_COMPUTE_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/arcmgmt/app/compute
+#EVENT_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/arcevent
+#AGENT_FWL_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-plugins
+PREBOOT_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/preboot
+#PREBOOT_COMPUTE_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/preboot/app/compute
+#PREBOOT_D1_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/preboot/app/preboot_d1
+AGENT_SERDES_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-plugins
+AGENT_PCIE_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-plugins
+#DSM_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/pcie-ss/dsm
+RESET_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/reset/app/io
+#RESET_COMPUTE_OVERRIDE_SRCDIR = $(TOPDIR)/../zephyr-apps/reset/app/compute
diff --git a/board/habanalabs/jgs/jgs-post-image-fit-arc.sh b/board/habanalabs/jgs/jgs-post-image-fit-arc.sh
new file mode 100755
index 0000000..3f0499d
--- /dev/null
+++ b/board/habanalabs/jgs/jgs-post-image-fit-arc.sh
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
+
+set -e          # stop on any error in this script
+set -o pipefail # stop on any error in pipeline
+
+source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/jgs-image-utility.sh
+DEV_NAME=`echo $HL_DEV | cut -d'_' -f1`
+BOARD_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
+MKIMAGE=${BR2_EXTERNAL}/output/${HL_DEV}/host/usr/bin/mkimage
+LZOP=${BR2_EXTERNAL}/output/${HL_DEV}/host/usr/bin/lzop
+
+# Set hl-version = "x.y.z"; inside FIT header to be used for KMD/FW compatability validation
+HL_VERSION=`cut -d- -f4 ${BR2_EXTERNAL}/output/${HL_DEV}/build/baremetal-custom/localversion-habanalabs`
+
+ZEPHYR_ITS_FILE=zephyr-${DEV_NAME}-${HL_TARGET}.its
+ZEPHYR_ITB_FILE=${ITBS_DIR}/zephyr-fit.itb
+
+# Set compression to lzo
+HL_COMPRESS=lzo
+
+source ${BOARD_BASE}/post-image-fit.sh
+source ${BOARD_BASE}/post-image-common.sh
+source ${BOARD_BASE}/post-image-exports.sh
+
+ZEPHYR_ITS_TEMPLATE=${BOARD_BASE}/jgs-zephyr.its
+
+ARC_MGMT_ICCM_EXT=${ARC_MGMT_ICCM_EXT:-iccm.bin.le}
+ARC_MGMT_DCCM_EXT=${ARC_MGMT_DCCM_EXT:-dccm.bin.le}
+
+ITS_SED_PARAMS+="s/@DEV_PREFIX@/${DEV_PREFIX}/g;"
+ITS_SED_PARAMS+="s/@HL_ARCH@/${HL_ARCH}/g;"
+ITS_SED_PARAMS+="s/@DEV_NAME@/${DEV_NAME}/g;"
+
+KEY_BASE=${BR2_EXTERNAL}/board/habanalabs/${DEV_PREFIX}/
+KEY_BASE_NAME=rsa3072
+KEY_NAME=hl_${KEY_BASE_NAME}
+
+ITS_SED_PARAMS+="s/@KEY_DTB@/${KEY_NAME}.dtb/g;"
+ITS_SED_PARAMS+="s/@KEY_EXT@/${KEY_NAME}_ext.bin.le/g;"
+
+# Replace ARC addresses
+ITS_SED_PARAMS+="s/@ARC_PID_DCCM_HI@/${ARC_PID_DCCM_HI}/g;"
+ITS_SED_PARAMS+="s/@ARC_PID_DCCM_LO@/${ARC_PID_DCCM_LO}/g;"
+ITS_SED_PARAMS+="s/@ARC_PID_ICCM_HI@/${ARC_PID_ICCM_HI}/g;"
+ITS_SED_PARAMS+="s/@ARC_PID_ICCM_LO@/${ARC_PID_ICCM_LO}/g;"
+ITS_SED_PARAMS+="s/@ARC_MGMT_DCCM_HI@/${ARC_MGMT_DCCM_HI}/g;"
+ITS_SED_PARAMS+="s/@ARC_MGMT_DCCM_LO@/${ARC_MGMT_DCCM_LO}/g;"
+ITS_SED_PARAMS+="s/@ARC_MGMT_ICCM_HI@/${ARC_MGMT_ICCM_HI}/g;"
+ITS_SED_PARAMS+="s/@ARC_MGMT_ICCM_LO@/${ARC_MGMT_ICCM_LO}/g;"
+
+# Replace ARC core IDs
+ITS_SED_PARAMS+="s/@FIT_CORE_ID_IO0_PID@/${FIT_CORE_ID_IO0_PID}/g;"
+ITS_SED_PARAMS+="s/@FIT_CORE_ID_IO0_MGMT@/${FIT_CORE_ID_IO0_MGMT}/g;"
+
+pushd ${STAGE_DIR}
+	hl_copy_keys ${BINS_DIR}
+
+	hl_copy_binaries ${BINS_DIR}
+
+	mkdir -p ${ITBS_DIR}
+
+	sed ${ITS_SED_PARAMS} ${ZEPHYR_ITS_TEMPLATE} > ${ZEPHYR_ITS_FILE}
+	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${ZEPHYR_ITS_FILE} -k ${STAGE_DIR} ${ZEPHYR_ITB_FILE}
+
+	dd status=none conv=sync bs=4 if=${ZEPHYR_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-boot-fit.itb
+popd
diff --git a/board/habanalabs/jgs/jgs-post-image-spi-arc_global.sh b/board/habanalabs/jgs/jgs-post-image-spi-arc_global.sh
new file mode 100755
index 0000000..3891514
--- /dev/null
+++ b/board/habanalabs/jgs/jgs-post-image-spi-arc_global.sh
@@ -0,0 +1,450 @@
+#!/usr/bin/env bash
+
+set -e          # stop on any error in this script
+set -u          # stop on any reference to undefine variable
+set -o pipefail # stop on any error in pipeline
+
+SHA256_SIZE=32
+SHA256_CHKSUM=sha256sum
+BOOTLOADER_BINFILE=${HL_TARGET}_bloader.bin
+COMP_BOOTLOADER_BINFILE=${HL_TARGET}_bloader_comp.bin
+HL_BASE_PATH=${BR2_EXTERNAL}/board/habanalabs
+BOARD_BASE=${HL_BASE_PATH}/${HL_DEV}
+DEVICE_BASE=${HL_BASE_PATH}/${HL_DEV}
+KEY_BASE=${BR2_EXTERNAL}/board/habanalabs/${DEV_PREFIX}
+SCRIPTS_DIR=${BR2_EXTERNAL}/scripts
+source ${DEVICE_BASE}/jgs-image-utility.sh
+OBJCOPY=objcopy
+ARCBOOT_BINFILE=${ARCBOOT_BINFILE:-boot_rom.bin}
+ARCBOOT_COMP_BINFILE=${ARCBOOT_COMP_BINFILE:-boot_rom_comp.bin}
+PPBOOT_BINFILE=${PPBOOT_BINFILE:-ppboot.bin}
+PREBOOT_BINFILE=${PREBOOT_BINFILE:-preboot.bin}
+#PREBOOT_COMPUTE_BINFILE=${PREBOOT_COMPUTE_BINFILE:-preboot_compute.bin}
+PCIE_PHY_BINFILE=${PCIE_PHY_BINFILE:-pcie_phy_fw.bin}
+
+#SECURITY_BOOT_MODE
+SECURE_BOOT=1
+
+#[REVOKE/HASH/FW_VERSION]_LOCATION
+ASIC_EFUSE=0
+
+#SPI_CONTROLLER
+SPI_CONT=1
+
+# current efuse settings
+REVOKE_LOCATION=${ASIC_EFUSE}
+HASH_LOCATION=${ASIC_EFUSE}
+SECURITY_BOOT_MODE=${SECURE_BOOT}
+ENABLE_SECURITY_VALIDATION=1
+DISBLE_DEBUG_IF=0
+FW_VERSION_LOCATION=${ASIC_EFUSE}
+SPI_CONTROLLER=${SPI_CONT}
+
+FLASH_BASE_ADDR=0x60000
+
+HABANA_KEY_REVOKE=0
+PROD1_KEY_REVOKE=0
+PROD2_KEY_REVOKE=0
+PROD3_KEY_REVOKE=0
+PROD4_KEY_REVOKE=0
+
+PRIME_IMAGE_PPBOOT_KEY=0
+PRIME_IMAGE_RAM_KEY=0
+
+PRIME_IMAGE_PPBOOT_CRC_ERROR=0
+PRIME_IMAGE_RAM_CRC_ERROR=0
+
+PRIME_IMAGE_PPBOOT_SIGN_ERROR=0
+PRIME_IMAGE_RAM_SIGN_ERROR=0
+
+KEY_BASE_NAME=rsa3072
+
+SECURITY_VER_FILE=${SPECS_EMB_PROJ_DIR}/${DEV_PREFIX}/${DEV_PREFIX}_security.h
+export BOOT_POINTER_HEADER=${SPECS_EMB_PROJ_DIR}/common/hl_boot_desc.h
+export FW_IF_HEADER=${SPECS_EMB_PROJ_DIR}/jgs/jgs_fw_if.h
+
+source ${DEVICE_BASE}/post-image-spi.sh
+
+SECURITY_FW_VERSION=$(grep -owE "HL_SEC_FW_VERSION[[:space:]]+[0-9]+" $SECURITY_VER_FILE | sed 's/^.*[ \t]\([0-9]\+\)/\1/;')
+ENDIANNESS="${ENDIANNESS:-be}"
+
+function hl_reverse_words {
+	FILE=$1
+	OUT_FILE=$2
+	WORD_SIZE=${3}
+
+	local filesize=$(wc -c < $FILE)
+	echo $FILE $filesize $WORD_SIZE
+	${OBJCOPY}  -Ibinary --reverse-bytes=${WORD_SIZE} ${FILE} ${OUT_FILE}
+}
+function hl_calc_crc32 {
+	FILE=$1
+
+	# reverse CRC32 source as DMA engine works like that
+	hl_reverse_words ${FILE} ${FILE}.out 4
+	CRC32=$(crc32 ${FILE}.out)
+	echo ${CRC32} | xxd -r -p > ${FILE}.crc32
+}
+
+function hl_sec_fw {
+	FILE=${1}
+	if [ ${2} -ge 64 ]; then
+		VERSION=$(printf "%016x" $(( ~$((1<<${2})) )))
+		VERSION+=$(printf "%016x" $(( ~0 )))
+	else
+		VERSION=$(printf "%016x" $(( ~0 )))
+		VERSION+=$(printf "%016x" $(( ~$((1<<${2})) )))
+	fi
+	echo ${VERSION} | xxd -r -p | xxd -e -g16 | xxd -r		> ${FILE}.tmp
+	dd status=none conv=notrunc seek=0 bs=16 if=${FILE}.tmp of=${FILE}
+
+	if [ ${HL_TARGET} == "palad" ]; then
+		if [ ${2} -ge 64 ]; then
+				VERSION=$(printf "%016x" $((1<<${2})))
+				VERSION+=$(printf "%016x" $((0)))
+			else
+				VERSION=$(printf "%016x" $((0)))
+				VERSION+=$(printf "%016x" $((1<<${2})))
+		fi
+		for i in {0..3}; do
+			DWORD=$(echo ${VERSION} | xxd -r -p | xxd -g4 | cut -d' ' -f$(( 5 - ${i} )))
+			FORCE="{h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_$(( 59 + ${i}))_r.data_in_r[31:0]}"
+			echo "force ${FORCE} 'h${DWORD}" >> ${HL_TARGET}-force.hex
+		done
+	fi
+}
+
+function hl_sec_key {
+	FILE=${1}
+	NAME=${2}
+        KEY_OFFSET=${3}
+        KEY_BASE_NAME_PUB=${KEY_BASE_NAME}.pub
+
+	calculate_hash ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB} ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash
+	cat ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash				> ${FILE}.tmp
+	dd status=none conv=notrunc seek=$(( ${KEY_OFFSET}*2 + 1 )) bs=16 if=${FILE}.tmp of=${FILE}
+
+	if [ ${HL_TARGET} == "palad" ]; then
+		for i in $(seq 1 $HASH_SZ); do
+			DWORD=$(xxd -e -c32 -g4 ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash | cut -d' ' -f$(( ${i} + 2 )))
+			OFFSET=$(( $(( ${KEY_OFFSET}*8 )) +  ${i} ))
+			FORCE="{h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_$(( 18 + ${OFFSET}))_r.data_in_r[31:0]}"
+			echo "force ${FORCE} 'h${DWORD}" >> ${STAGE_DIR}/${HL_TARGET}-force.hex
+		done
+	fi
+}
+
+function hl_key_revoke {
+	REVOKE_STATUS=$1
+	if [ $REVOKE_STATUS -eq 1 ]; then
+		echo -n fe >> ${STAGE_DIR}/revoke_status.hex
+	else
+		echo -n ff >> ${STAGE_DIR}/revoke_status.hex
+	fi
+}
+
+function hl_efuse_sys_config_generate {
+	DWORD=${HABANA_KEY_REVOKE}
+	DWORD=$(( DWORD + $((${PROD1_KEY_REVOKE}<<1))))
+	DWORD=$(( DWORD + $((${PROD2_KEY_REVOKE}<<2))))
+	DWORD=$(( DWORD + $((${PROD3_KEY_REVOKE}<<3))))
+	DWORD=$(( DWORD + $((${PROD4_KEY_REVOKE}<<4))))
+	DWORD=$(( DWORD + $((${REVOKE_LOCATION}<<5))))
+	DWORD=$(( DWORD + $((${HASH_LOCATION}<<7))))
+	DWORD=$(( DWORD + $((${SECURITY_BOOT_MODE}<<9))))
+	DWORD=$(( DWORD + $((${ENABLE_SECURITY_VALIDATION}<<10))))
+	DWORD=$(( DWORD + $((${DISBLE_DEBUG_IF}<<12))))
+	DWORD=$(( DWORD + $((${FW_VERSION_LOCATION}<<13))))
+	DWORD=$(( DWORD + $((${SPI_CONTROLLER}<<17))))
+	DWORD=$(printf "%08x" ${DWORD})
+	echo "force {h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_58_r.data_in_r[31:0]} 'h${DWORD}" > ${HEXS_DIR}${HL_TARGET}-force.hex
+}
+
+function hl_report_missing_spi_region_offset_and_exit {
+	local red=$(tput setaf 1)
+	local normal=$(tput sgr0)
+
+	printf "$red The offset of SPI flash region %s is not defined $normal \n" $1
+	exit 1
+}
+
+function hl_report_missing_spi_region_size_and_exit {
+	local red=$(tput setaf 1)
+	local normal=$(tput sgr0)
+
+	printf "$red The size of SPI flash region %s is not defined $normal \n" $1
+	exit 1
+}
+
+# This function checks, that for each SPI flash region
+# both offset and size are defined in jgs_mem_addr_dt.h
+#
+# Parameters:
+# 1 - associative array region name to file name
+# 2 - associative array region name to offset
+# 3 - associative array region name to size
+# Result:
+# normal return if everything is OK
+# printing error message and exit if problem encountered
+function hl_validate_spi_flash_layout {
+	local -n spi_region_to_file=$1
+	local -n spi_region_offsets=$2
+	local -n spi_region_sizes=$3
+	local region_name
+
+	for region_name in "${!spi_region_to_file[@]}"; do
+		if [[ -z "${spi_region_offsets[$region_name]+unset}" ]]; then
+			hl_report_missing_spi_region_offset_and_exit $region_name
+		fi
+		if [[ -z "${spi_region_sizes[$region_name]+unset}" ]]; then
+			hl_report_missing_spi_region_size_and_exit $region_name
+		fi
+	done
+}
+
+# This function parses header file spec/jgs_mem_addr_dt.h
+# and learns SPI flash layout: region names, their offsets
+# and sizes
+# Result:
+# function fills 2 associaive arrays:
+# region name to offset and region name to size
+function hl_parse_spi_flash_regions {
+	local -n spi_region_offsets=$1
+	local -n spi_region_sizes=$2
+	local platform=$3
+	local mem_addr_header_file=${BR2_EXTERNAL}/../specs/fs/${platform}/${platform}_mem_addr_dt.h
+	local regex_offset="\#define(.*)SPI_FLASH_REGION_([A-Z_0-9]+)_OFFSET(\s+)([0-9a-fA-F]+)"
+	local regex_size="\#define(.*)SPI_FLASH_REGION_([A-Z_0-9]+)_SIZE(\s+)([0-9a-fA-F]+)"
+	local region_name
+	local region_offset
+	local region_size
+	local red=$(tput setaf 1)
+	local normal=$(tput sgr0)
+
+	if [ ! -f $mem_addr_header_file ]; then
+		printf "$red File %s not exists $normal \n" "$mem_addr_header_file"
+		exit 1
+	fi
+	IFS=$'\n'
+	for line in $(cat "$mem_addr_header_file"); do
+		if [[ $line =~ $regex_offset ]]; then
+			region_name=${BASH_REMATCH[2]}
+			region_offset=$(echo "ibase=16; ${BASH_REMATCH[4]^^}"|bc)
+			spi_region_offsets[${region_name}]=$region_offset
+		elif [[ $line =~ $regex_size ]]; then
+			region_name=${BASH_REMATCH[2]}
+			region_size=$(echo "ibase=16; ${BASH_REMATCH[4]^^}"|bc)
+			spi_region_sizes[${region_name}]=$region_size
+		fi
+	done
+}
+
+# This function is intended to check, that the length of
+# given files does not exceed the size of appropriate SPI
+# flash region
+function hl_check_length_of_spi_files {
+	local -n region_to_file=$1
+	local -n region_to_size=$2
+	local file_name
+	local region_name
+	local block_size
+	local diff
+	local red=$(tput setaf 1)
+	local normal=$(tput sgr0)
+
+	for region_name in "${!region_to_file[@]}"; do
+		file_name="${region_to_file[${region_name}]}"
+		[[ $file_name == FF ]] && continue
+		[[ $file_name == 00 ]] && continue
+		block_size="${region_to_size[${region_name}]}"
+		if [ ! -f $file_name ]; then
+			printf "$red File %s not exists $normal \n" "$file_name"
+			exit 1
+		fi
+		file_size=$(wc -c <"$file_name")
+		if [ $file_size -eq 0 ]; then
+			printf "$red File %s is empty $normal \n" "$file_name"
+			exit 1
+		fi
+		if [ $file_size -gt $block_size ]; then
+			diff=$(expr $file_size - $block_size)
+			printf "$red File %s size %s exceeds SPI block size %s by %s bytes $normal \n" \
+				"$file_name" "$file_size" "$block_size" "$diff"
+			exit 1
+		else
+			printf "Region %s file size: %s    SPI block size: %s\n" \
+				"$region_name" "$file_size" "$block_size"
+		fi
+	done
+}
+
+# This function builds SPI flash file
+function hl_build_spi_flash_file {
+	local -n region_to_file=$1
+	local -n region_to_offset=$2
+	local -n region_to_size=$3
+	local -n out_file=$4
+	local file_length=0
+	local input_file
+	local file_length
+	local offset
+	local size
+	local kb
+	local end_of_region
+	local current_region_offset=0
+	local region_found
+	local red=$(tput setaf 1)
+	local normal=$(tput sgr0)
+
+	for region_name in "${!region_to_offset[@]}"; do
+		offset="${region_to_offset[${region_name}]}"
+		size="${region_to_size[${region_name}]}"
+		end_of_region=$(expr $offset + $size)
+		if [ $end_of_region -gt $file_length ]; then
+			file_length=$end_of_region
+		fi
+	done
+
+	printf "\nLength of flash image file %s: 0x%x bytes \n\n" "$out_file" "$file_length"
+	rm -f "$out_file"
+	touch "$out_file"
+
+	while [ $current_region_offset -lt $file_length ]; do
+		region_found=false
+		for region_name in "${!region_to_offset[@]}"; do
+			offset="${region_to_offset[${region_name}]}"
+			size="${region_to_size[${region_name}]}"
+			kb=$(expr "$size" \/ 1024)
+			if [ $offset -eq $current_region_offset ]; then
+				region_found=true
+				input_file="${region_to_file[${region_name}]}"
+				if [ "$input_file" = "FF" ]; then
+					printf "Writing FF to offset 0x%x  %sK (Region name: %s)\n" "$offset" "$kb" "$region_name"
+					dd status=none conv=sync bs=${kb}K count=1 if=/dev/zero | tr "\000" "\377" >> ${out_file}
+				elif [ "$input_file" = "00" ]; then
+					printf "Writing 00 to offset 0x%x  %sK (Region name: %s)\n" "$offset" "$kb" "$region_name"
+					dd status=none conv=sync bs=${kb}K count=1 if=/dev/zero >> ${out_file}
+				else
+					printf "Writing %s from %s to offset 0x%x  %sK \n" "$region_name" "$input_file" "$offset" "$kb"
+					dd status=none conv=sync bs=${kb}K if=${input_file} >> ${out_file}
+				fi
+				current_region_offset=$(expr $offset + $size)
+				break
+			fi
+		done
+		if [ "$region_found" = false ]; then
+			printf "$red SPI flash region with offset 0x%x not defined $normal \n" "$current_region_offset"
+			exit 1
+		fi
+	done
+}
+
+pushd ${STAGE_DIR}
+
+	hl_copy_binaries ${BINS_DIR}
+
+	export IMG_PTR_SZ="$(get_struct_size $BOOT_POINTER_HEADER hl_fw_boot_ptrs)"
+	export IMG_DESC_SZ="$(get_struct_size $BOOT_POINTER_HEADER hl_fw_img_desc)"
+
+	for i in $(seq 1 $NUM_PROD_KEYS); do
+		CUST=cust${i}
+		hl_gen_cert ${CUST} "Habana Embedded Team"
+	done
+
+	## create otp spi image
+	dd status=none conv=sync count=12 bs=16 if=/dev/zero | tr '\000' $'\xFF' > ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin
+
+	# get private and public keys
+	copy_all_keys ${BINS_PROD_DIR}
+
+	#generate public key components
+	generate_public_keys
+
+	# Create erom binaries
+	prep_image_erom ${BINS_DIR}/${DEV_PREFIX}-erom.bin.le
+
+	# Create fake fuse binary
+	prep_image_fake_fuse ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le
+
+	# Pad preboot binary and then pack
+	zephyr_preboot_pack ${BINS_DIR}/${PREBOOT_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+	#zephyr_preboot_pack ${BINS_DIR}/${PREBOOT_COMPUTE_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+
+	# create combined name of ppboot and pcie
+	PPBOOT=$(echo "$PPBOOT_BINFILE" | cut -f 1 -d '.')
+	PCIE_PHY=$(echo "$PCIE_PHY_BINFILE" | cut -f 1 -d '.')
+	PPBOOT_PCIE_PHY_BINFILE=${PPBOOT}-${PCIE_PHY}.bin
+	PPBOOT_PCIE_PHY_PADDED_BINFILE=${PPBOOT_PCIE_PHY_BINFILE}.pad
+
+	# concatenate both FW
+	hl_join_bin ${BINS_DIR}/${PPBOOT} ${PCIE_PHY} ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
+
+	# pack and sign image
+	hl_pack_bin ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE} 64K ${PRIME_IMAGE_PPBOOT_KEY} ${PRIME_IMAGE_PPBOOT_CRC_ERROR} ${PRIME_IMAGE_PPBOOT_SIGN_ERROR}
+
+	# Create image pointers
+	prep_image_ptrs ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le
+
+	if [ ${HL_TARGET} == "palad" ]; then
+		hl_efuse_sys_config_generate
+	fi
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.le ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le  ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.le ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.be
+	#swap_file ${BINS_DIR}/${DEV_PREFIX}-preboot_compute.bin.le ${BINS_DIR}/${DEV_PREFIX}-preboot_compute.bin.be
+
+	# ToDo change it to unified wat from agent fw it check FSW-3432
+	create_spi_image ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc \
+			 ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.${ENDIANNESS}
+
+	# SPI flash binary is provided as Big Endian, for manufacturing programmer.
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le
+
+	# Prepare checksum
+	CKSUM=`${SHA256_CHKSUM} ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le | cut -d' ' -f1`
+	echo ${CKSUM} | xxd -p -r								> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.crc
+	dd status=none conv=sync bs=64K			if=${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le	> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin
+	dd status=none conv=sync bs=$SHA256_SIZE	if=${BINS_DIR}/${DEV_PREFIX}-spi-flash.crc		>> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin
+
+	rm -rf ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
+
+	if [ ${HL_TARGET} == "palad" ]; then
+		# convert io die bootrom to pallad hex image
+		hl_pad_arcboot_rom
+		${SCRIPTS_DIR}/palad/convert_app_to_io_bootrom_hex.sh ${ARCBOOT_BINFILE}
+		mv boot_rom_rom_ecc.hex ${HEXS_DIR}
+
+		# convert compute die bootrom to pallad hex images
+		${SCRIPTS_DIR}/palad/convert_app_to_compute_bootrom_hex.sh ${ARCBOOT_COMP_BINFILE} fs1-${HL_TARGET}-bootrom_comp
+		mv boot_rom_comp_*.hex ${HEXS_DIR}
+		mv jgs-${HL_TARGET}-bootrom_comp_*_ecc.hex ${HEXS_DIR}
+
+		# convert io die bootloader to pallad hex image
+		${SCRIPTS_DIR}/palad/convert_app_to_io_bootrom_hex.sh ${BOOTLOADER_BINFILE}
+
+		# convert compute die bootloader to pallad hex images
+		${SCRIPTS_DIR}/palad/convert_app_to_compute_bootrom_hex.sh ${COMP_BOOTLOADER_BINFILE} fs1-${HL_TARGET}-bloader_comp
+		mv jgs-${HL_TARGET}-bloader*.hex ${HEXS_DIR}
+
+		# convert preboot application to pallad hex images
+		${SCRIPTS_DIR}/palad/convert_app_to_ccm_hex.sh preboot.bin
+		mv iccm0_*.hex ${HEXS_DIR}
+
+		# convert preboot comp application to palad hex images
+		${SCRIPTS_DIR}/palad/convert_app_to_ccm_hex.sh preboot_compute.bin
+		mv iccm0_0.hex ${HEXS_DIR}/comp_iccm0_0.hex
+		mv iccm0_1.hex ${HEXS_DIR}/comp_iccm0_1.hex
+	fi
+
+	if [ -f "${KEY_BASE}/spi-flash-layout-validator.sh" ]; then
+		source ${KEY_BASE}/spi-flash-layout-validator.sh
+		validate_spi_flash_layout ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.${ENDIANNESS} \
+					  ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.${ENDIANNESS} \
+					  ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.${ENDIANNESS} \
+					  ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.${ENDIANNESS}
+	fi
+popd
diff --git a/board/habanalabs/jgs/jgs-zephyr.its b/board/habanalabs/jgs/jgs-zephyr.its
new file mode 100644
index 0000000..1f935e2
--- /dev/null
+++ b/board/habanalabs/jgs/jgs-zephyr.its
@@ -0,0 +1,95 @@
+/dts-v1/;
+
+/ {
+	hl-version = "@HL_VERSION@";
+	hl-device = "@DEV_PREFIX@";
+	description = "FIT image with Zephyr loadables";
+	#address-cells = <2>;
+
+	images {
+
+		/* TEMPORARY:
+		 * arcpid implementation isn't ready yet, so comment out
+		 * zephyr-pid-iccm & zephyr-pid-dccm.
+		pid-io-iccm {
+			description = "Zephyr pid ICCM";
+			data = /incbin/("@BINS_DIR_PATH@@DEV_NAME@-zephyr-pid-iccm.bin.le");
+			type = "standalone";
+			compression = "none";
+			loadaddr = <@ARC_PID_ICCM_HI@ @ARC_PID_ICCM_LO@>;
+			hash {
+				algo = "@SHA_VERSION@";
+			};
+		};
+
+		pid-io-dccm {
+			description = "Zephyr pid DCCM";
+			data = /incbin/("@BINS_DIR_PATH@@DEV_NAME@-zephyr-pid-dccm.bin.le");
+			type = "standalone";
+			compression = "none";
+			loadaddr = <@ARC_PID_DCCM_HI@ @ARC_PID_DCCM_LO@>;
+			hash {
+				algo = "@SHA_VERSION@";
+			};
+		};
+		*/
+
+		mgmt-io-iccm {
+			description = "Zephyr mgmt io ICCM";
+			data = /incbin/("@BINS_DIR_PATH@@DEV_NAME@-zephyr-mgmt-iccm.bin.le");
+			type = "standalone";
+			compression = "none";
+			instances = <1>;
+			loadaddr = <@ARC_MGMT_ICCM_HI@ @ARC_MGMT_ICCM_LO@>;
+			hash {
+				algo = "@SHA_VERSION@";
+			};
+		};
+
+		mgmt-io-dccm {
+			description = "Zephyr mgmt io DCCM";
+			data = /incbin/("@BINS_DIR_PATH@@DEV_NAME@-zephyr-mgmt-dccm.bin.le");
+			type = "standalone";
+			compression = "none";
+			instances = <1>;
+			loadaddr = <@ARC_MGMT_DCCM_HI@ @ARC_MGMT_DCCM_LO@>;
+			hash {
+				algo = "@SHA_VERSION@";
+			};
+		};
+	};
+
+	cpus {
+		arc-io-mgmt {
+			arch = "arc";
+			cpubase = <@MGMT_IO_BASEADDR_HI@ @MGMT_IO_BASEADDR_LO@>;
+			cputype = <0>; /* 0 - PSOC ARC, 1 - ARCFARM */
+			memnum = <2>;
+			membase = <@ARC_MGMT_ICCM_HI@ @ARC_MGMT_ICCM_LO@ @ARC_MGMT_DCCM_HI@ @ARC_MGMT_DCCM_LO@>;
+			memsize = <@MGMT_IO_ICCM_SIZE@ @MGMT_IO_DCCM_SIZE@>;
+			startaddr = <@MGMT_STARTADDR_HI@ @MGMT_STARTADDR_LO@>;
+		};
+
+		arc-io-pid {
+			arch = "arc";
+			cpubase = <@PID_IO_BASEADDR_HI@ @PID_IO_BASEADDR_LO@>;
+			cputype = <0>; /* 0 - PSOC ARC, 1 - ARCFARM */
+			memnum = <2>;
+			membase = <@ARC_PID_ICCM_HI@ @ARC_PID_ICCM_LO@ @ARC_PID_DCCM_HI@ @ARC_PID_DCCM_LO@>;
+			memsize = <@PID_IO_ICCM_SIZE@ @PID_IO_DCCM_SIZE@>;
+			startaddr = <@PID_STARTADDR_HI@ @PID_STARTADDR_LO@>;
+		};
+	};
+
+	configurations {
+		default = "preboot-conf";
+
+		preboot-conf {
+			load-type = <0>; /* FIT_LOAD_TYPE_RUN */
+			description = "Habanalabs device preboot";
+			loadables = /* "pid-io-iccm", "pid-io-dccm", */ "mgmt-io-iccm",
+				    "mgmt-io-dccm";
+			processors = "arc-io-mgmt"; /*, "arc-io-pid";*/
+		};
+	};
+};
diff --git a/board/habanalabs/jgs/key/hl_rsa3072.crt b/board/habanalabs/jgs/key/hl_rsa3072.crt
new file mode 100644
index 0000000..7a35572
--- /dev/null
+++ b/board/habanalabs/jgs/key/hl_rsa3072.crt
@@ -0,0 +1,26 @@
+-----BEGIN CERTIFICATE-----
+MIIEYDCCAsigAwIBAgIJAIdm15P2CRx0MA0GCSqGSIb3DQEBCwUAMEUxCzAJBgNV
+BAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBX
+aWRnaXRzIFB0eSBMdGQwHhcNMjEwMjE3MTQ1NzI3WhcNMjEwMzE5MTQ1NzI3WjBF
+MQswCQYDVQQGEwJBVTETMBEGA1UECAwKU29tZS1TdGF0ZTEhMB8GA1UECgwYSW50
+ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIB
+igKCAYEAobJwKRTZU0O848ir8SBHKAxvASUu80aYL/aBNbXnIAmxwZakiuxQKqS/
+cJ90OXRINYwtQL2dgjvEvc8z67juYJUFBEv6m8ZA6a6n3fJ96IlLV58rJhpcHngj
+b+vbP9pgD2NV07FMOP8wEs1G6GwJUrePoif3+U1aOJeiWtgvoz6n+rYmiwLaADJo
+DHtBotBsNGLM6Zs9Ta4jnmRA+Ln6dQgJnRGkMoPSOoV1RFIlfCVIMXu3LvkhzYh6
+2eQs3ASa7vJ/LvX4MUwTXk3pd+blxkwo06fQxbHR6aMIvZZ2z5bQsyPJjqziwufc
+OlFymAUP5KlCmnSyKnmVST109g+vuwUiGwRpr0MvCpuo2i0oq90NpETi2bHDUZLN
+JP/UmZWW9FuFungkjeNSjpqoV6sxXk5RhK5NKU3n3oQslL2YjaZ57WHa+oyEnCew
+ZQCrD9H9z+8cUdsvHd6tyEkjSR6boWbVkPnoQ6QMDGcFcy5gzDzn/CaaFSBSvTNH
+K4VHjUk1AgMBAAGjUzBRMB0GA1UdDgQWBBSSNuBPFHgSUGLde2uMlRUw10/vhjAf
+BgNVHSMEGDAWgBSSNuBPFHgSUGLde2uMlRUw10/vhjAPBgNVHRMBAf8EBTADAQH/
+MA0GCSqGSIb3DQEBCwUAA4IBgQBUVNhiV5Yj0fU471SyCIv+C3FbU0umDcBTxECc
+2RQRTtwiT3rUrDnkuM9dpqUdUfltw7nB73m2gG7SjWd+tmTnJZQCss53xzfOoLfm
+Wn8j65prNScVWZaDH3vmwV6puPBakF0ihMOkPIxPsZcFZ6sPLUscHpL0Ohq/gi1M
+N25WGzdaksWhoHM08y8WrAVJj5iSyBZ5Pw0WMYKjgljvJYwrV36I7x8oUX6bQ8z0
+DLSd4HmU210YEBhmz+/r01toMAilW3bdOLERL1L+7bYAnoxnz+HPFU+HNluXxbCU
+1SsRLgKSJO2O17BVpJ2iWkAWkNb6Cc2PW5hbTju0cx89fmR5UbTnBZwFBinARfqX
+8TN0Ws9ayVedz8bMtefFgGgTYwmz+cvMLTp61LNzAXKbkhAam/gyhCXknsqwPQEB
+Jyb1Crg5wwKBDGVSu/io0koabYOykvBKOJLtDQOYJHs2bTCzEXxuERnAWaNuLbDV
+T0PW0+uH18tgCTDdzaimymLFIzg=
+-----END CERTIFICATE-----
diff --git a/board/habanalabs/jgs/key/hl_rsa3072.dtb b/board/habanalabs/jgs/key/hl_rsa3072.dtb
new file mode 100644
index 0000000..3ed808e
--- /dev/null
+++ b/board/habanalabs/jgs/key/hl_rsa3072.dtb
Binary files differ
diff --git a/board/habanalabs/jgs/key/hl_rsa3072.key b/board/habanalabs/jgs/key/hl_rsa3072.key
new file mode 100644
index 0000000..018862e
--- /dev/null
+++ b/board/habanalabs/jgs/key/hl_rsa3072.key
@@ -0,0 +1,40 @@
+-----BEGIN PRIVATE KEY-----
+MIIG/QIBADANBgkqhkiG9w0BAQEFAASCBucwggbjAgEAAoIBgQChsnApFNlTQ7zj
+yKvxIEcoDG8BJS7zRpgv9oE1tecgCbHBlqSK7FAqpL9wn3Q5dEg1jC1AvZ2CO8S9
+zzPruO5glQUES/qbxkDprqfd8n3oiUtXnysmGlweeCNv69s/2mAPY1XTsUw4/zAS
+zUbobAlSt4+iJ/f5TVo4l6Ja2C+jPqf6tiaLAtoAMmgMe0Gi0Gw0Yszpmz1NriOe
+ZED4ufp1CAmdEaQyg9I6hXVEUiV8JUgxe7cu+SHNiHrZ5CzcBJru8n8u9fgxTBNe
+Tel35uXGTCjTp9DFsdHpowi9lnbPltCzI8mOrOLC59w6UXKYBQ/kqUKadLIqeZVJ
+PXT2D6+7BSIbBGmvQy8Km6jaLSir3Q2kROLZscNRks0k/9SZlZb0W4W6eCSN41KO
+mqhXqzFeTlGErk0pTefehCyUvZiNpnntYdr6jIScJ7BlAKsP0f3P7xxR2y8d3q3I
+SSNJHpuhZtWQ+ehDpAwMZwVzLmDMPOf8JpoVIFK9M0crhUeNSTUCAwEAAQKCAYBI
+Y7ZK7gaNfFnoMEKp5KV2jHp1rMvkJcqkTltyTEl40l4+KM7pUvpg3PjH1tsVjbV2
+I5du0HTfhRqDT2wTL+uoWY6xacJ8Urlk5MhKWPF9VVnhlGSg3lQ3X9mdlZ3GbGFt
+Ai++YjWZnsVK5ZPFIfjai0tGcVZWygaGJ3yL5xUMpFRXrYx71375b5edXuD3nw+S
+NFlAjDtZt4hCtOdRcxFpq3uCUg+Mfe90O6NQPuaBOb+kR5qu8OHtPhihy8E5juWH
+kRF3mzjvtu1rNE1Xngctg0MnIkrEqhgYfSpVtZD7MOoVXVSmtnizB8u3Nj5cG902
+d9yEw33ifVq9tSwddbLnw3YOMonqffQQRTXBxcutACV3MWK6ExKYvPU5D3sxKgJC
+kiuFy1pL43XID3cXKRh23CzNESxyyrq0/o+pglgZTtadsQkvjqoMbNDOZ7xJUlMh
+b6iACn3PphLXVQDI7QFssea8hnoro9RbHAbD5MVadXLTPqWpjWAOtFKPs62x8NEC
+gcEA1Myg6vmZQ4I2t0RErEhTAWmD1KZJdSZF2GWuZ09PwW7Wl1QWUORo//Nb520v
+5mtXWdlt+pYZD4Mnf5oW3BSIsGdqf+0CVBsAA9awlZXhOvzk29Ph05MPjfUPOJdr
+m5A6Ny0ZOPBw6jZhS8hMa9nAT7aKjNX4bmQXCfxZUS4HKrzDAvqXhTFhS98lSQ/q
+E2NX8CPX8BIIVBfa0ge81RL0WkLUW0OEDVeIG5f7g4RsV6O8f47zEM2JVLKiuRrF
+nOSPAoHBAMKF+hxd8yDij+Lgzfq90gOU2BIMpTvpNxhrFWrEClM1geyPbfZQLSzg
+MyGPKGSE8XIXqn5LZZLsrkIeIkr3ajakBLkQ+x5cO+K0lV3u3ktS0rAFSEMKqb/H
+65dq2dtjdfesA82FDGZN0nt9hZH7Hew+m5M+H9rsxO5qiZZCUU+WRDnOed05IVUK
+Mv2SW43Xsk0UgfwygIAZ79kLaDdoFgoRypF97slADkNhrVgXmnyrws/jfBYzgFMR
+49PFTuY/+wKBwGQ4Dx06+fx6Seyb4T/GPBlxKR4wRm4+3SwiDzYmeyhdXAEfH/du
+6yiH42Ol36ZELXpqVUpsAFd27FZDKrrxjMKPoe+nnq4fnS0xoi6AzPdr/RiuvC9Z
+ljzECiwb3BkCnvg3bCaNzo/vlnkLVzp/7ftcu/nCF+NdjiAs0Z/S7pspFSrE/DFF
+ZU5HRJcZQKNq6RJ2vw8YXwXDylGaRUuQRrZvvPVSYdGqxLZCKLjy19eajH8CwuTJ
++ACEtPA6dJEEbwKBwDOM/fN2knAsmqUtnmjM1VNNptVhtIhkDc3S0yIwQew7Ses/
+N40HQnGaGPLT1YiOEs6BpxVBNfHx91xeUAg9+U/xt66z6VjqaJRz9wGlxBOCJ4Ou
+YrAhq0q/2JeQIZwah7DTTU6x689Q4G8JnNRLEGsDTIeA90A1bxsf8MZ5XMMSt5H6
+PiF6enLm+QxqNQkqGH6E6M3FNHi+ZunGNCA63Vxdyqk0DowXBbCluiKWXsKbYk1l
+pATeMSvSmeYNbBNo8QKBwQCyxs/rQv/6uqPr5VkAp6sfFlo0eku9BYIgjdfN+vvw
+Z+qR/O0j3e1s3MgIhpuULswrsMOPu8W1Y+B0k68VHCZ0oEbMHwSve+qDO5ozWwGj
+stTUVl+7LWsfHZY3p1bIX61sBMPUvnwqDvhbq6L9vWfh6ZGXcYhEEYzuYZua00W8
+CSH7VChITTUqLAK5y9KDagfSvgGb2Q6JFhNexDgIm2tXgl7cj3Z8aCY4VztgmLBV
+l4m0hTJutcL9fA2WvkHLKgU=
+-----END PRIVATE KEY-----
diff --git a/board/habanalabs/jgs/post-image-agents.sh b/board/habanalabs/jgs/post-image-agents.sh
new file mode 100755
index 0000000..be740f0
--- /dev/null
+++ b/board/habanalabs/jgs/post-image-agents.sh
@@ -0,0 +1,85 @@
+# !/usr/bin/env bash
+
+# TODO: security and agent image related lines commented out intentionally and will be supported later on.
+
+set -e  	# stop on any error in this script
+set -u 		# stop on any reference to undefine variable
+set -o pipefail # stop on any error in pipeline
+
+source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/post-image-spi.sh
+#source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/create-fake-fuse-image.sh
+
+DEV_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
+source ${DEV_BASE}/jgs-image-utility.sh
+
+hl_create_section_variables
+
+HL_COMPRESS=none
+HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/preboot-custom/build/buildversion-habanalabs`
+HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/baremetal-custom/buildversion-habanalabs | cut -d ' ' -f1`
+HL_FW_VERSION=""
+AGENT_BASE_ADDR=0x0000000000000000
+#SECURITY_SPI_OFFSET=0xFFFFFFFF
+ENDIANNESS="be"
+AGENT_ARCH="arc"
+
+LOADABLES="ppboot\", \"preboot\", \"img-ptr"
+
+# Extract the load address and start address from stat file
+ITS_SED_PARAMS+="s/@HL_PREBOOT_VERSION_SECUREBOOT@/${HL_PREBOOT_VERSION_SECUREBOOT}/g;"
+ITS_SED_PARAMS+="s/@HL_PPBOOT_VERSION_SECUREBOOT@/${HL_PPBOOT_VERSION_SECUREBOOT}/g;"
+ITS_SED_PARAMS+="s|@HL_DEV@|${DEV_PREFIX}|g;"
+ITS_SED_PARAMS+="s/@ENDIANNESS@/${ENDIANNESS}/g;"
+ITS_SED_PARAMS+="s/@AGENT_ARCH@/${AGENT_ARCH}/g;"
+#ITS_SED_PARAMS+="s/@SEC_OFFSET@/${SECURITY_SPI_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@IMG_PTR_PRI_OFFSET@/0x${SPI_FLASH_REGION_IMAGE_DESC_PRIMARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@IMG_PTR_SEC_OFFSET@/0x${SPI_FLASH_REGION_IMAGE_DESC_SECONDARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PPBOOT_PRI_OFFSET@/0x${SPI_FLASH_REGION_PPBOOT_PRIMARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PPBOOT_SEC_OFFSET@/0x${SPI_FLASH_REGION_PPBOOT_SECONDARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_PRI_OFFSET@/0x${SPI_FLASH_REGION_PREBOOT_IO_DIE_PRIMARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_SEC_OFFSET@/0x${SPI_FLASH_REGION_PREBOOT_IO_DIE_SECONDARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_COMPUTE_PRI_OFFSET@/0x${SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_PRIMARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_COMPUTE_SEC_OFFSET@/0x${SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_SECONDARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@FAKE_FUSE_PRI_OFFSET@/0x${SPI_FLASH_REGION_FAKE_FUZE_PRIMARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@FAKE_FUSE_SEC_OFFSET@/0x${SPI_FLASH_REGION_FAKE_FUZE_SECONDARY_OFFSET}/g;"
+ITS_SED_PARAMS+="s/@IMG_BIN@/images-pointers/g;"
+ITS_SED_PARAMS+="s/@PPBOOT_BIN@/ppboot-pcie_phy_fw/g;"
+ITS_SED_PARAMS+="s/@FAKE_FUSE_BIN@/fake-fuse/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_BIN@/preboot/g;"
+ITS_SED_PARAMS+="s/@PREBOOT_COMPUTE_BIN@/preboot_compute/g;"
+ITS_SED_PARAMS+="s/@FW_LOADER_CONF_LOADABLES@/${LOADABLES}/g;"
+
+#AGENT_NIC_FW_FILE=${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_NAME}-eth-phy.bin.le
+#AGENT_SERDES_FW_NAME=zephyr_agentfw_serdes.bin
+
+#source ${DEV_BASE}/jgs-image-utility.sh
+hl_create_section_variables
+
+#ITS_SED_PARAMS+="s/@AGENT_FW_ADDR_HI@/${NIC_PHY_FW_ADDR_HI}/g;"
+#ITS_SED_PARAMS+="s/@AGENT_FW_ADDR_LO@/${NIC_PHY_FW_ADDR_LO}/g;"
+
+function handle_fw_copy {
+	# Set NIC FW to be the serdes agent FW file
+	#cp ${AGENT_NIC_FW_FILE} ./${AGENT_SERDES_FW_NAME}
+
+	# Copy common arc files required for agent fit creation
+	handle_fw_copy_arc
+}
+
+function handle_preboot_loadaddr {
+	return
+}
+
+function post_image_agent_handle {
+	return
+}
+
+function  handle_secondary_agent_fit {
+	# Remove security node
+	fdtput -r ${ITBS_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE} /images/security
+	# Add core-id property for agent
+	fdtput -p -tx ${ITBS_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE} /images/agent core-id 0x1
+	dd status=none conv=sync bs=4 if=${ITBS_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE} > ${STAGE_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE}.tmp
+	mv ${STAGE_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE}.tmp ${ITBS_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE}
+	handle_agent_fit erom ${COMMON_BASE}
+}
diff --git a/board/habanalabs/jgs/post-image-common.sh b/board/habanalabs/jgs/post-image-common.sh
new file mode 100755
index 0000000..811e3ef
--- /dev/null
+++ b/board/habanalabs/jgs/post-image-common.sh
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+source ${BR2_EXTERNAL}/board/habanalabs/common/dirs.sh
+
+SHA_VERSION=sha384
+BIN_IMG_DIR=${BR2_EXTERNAL}/output/${HL_DEV}_arc/images
+KEYS_DIR=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}_arc/key/
+
+# Set enum number of CORE IDs in FIT
+FIT_CORE_ID_IO0_PID=$(grep -E "FIT_CORE_ID_IO0_PID[^_]" ${SPECS_EMBEDDED_DIR}/fs/common/hl_fit_if.h | sed 's/.* = \(.*\),/\1/')
+FIT_CORE_ID_IO0_MGMT=$(grep -E "FIT_CORE_ID_IO0_MGMT[^_]" ${SPECS_EMBEDDED_DIR}/fs/common/hl_fit_if.h | sed 's/.* = \(.*\),/\1/')
+FIT_CORE_ID_IO0_RST=$(grep -E "FIT_CORE_ID_IO0_RST[^_]" ${SPECS_EMBEDDED_DIR}/fs/common/hl_fit_if.h | sed 's/.* = \(.*\),/\1/')
diff --git a/board/habanalabs/jgs/post-image-exports.sh b/board/habanalabs/jgs/post-image-exports.sh
new file mode 100755
index 0000000..38c21fc
--- /dev/null
+++ b/board/habanalabs/jgs/post-image-exports.sh
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+ITS_SED_PARAMS+="s/@HL_VERSION@/${HL_VERSION}/g;"
+ITS_SED_PARAMS+="s/@HL_DEV@/${HL_DEV}/g;"
+ITS_SED_PARAMS+="s/@HL_PREBOOT_VERSION@/${HL_PREBOOT_VERSION}/g;"
+ITS_SED_PARAMS+="s/@HL_TARGET@/${HL_TARGET}/g;"
+ITS_SED_PARAMS+="s/@FIT_LOAD_TYPE_FLASH@/${FIT_LOAD_TYPE_FLASH}/g;"
+ITS_SED_PARAMS+="s/@FIT_LOAD_TYPE_RUN@/${FIT_LOAD_TYPE_RUN}/g;"
+ITS_SED_PARAMS+="s/@HL_FW_VERSION@/${HL_FW_VERSION}/g;"
+#ITS_SED_PARAMS+="s/@HL_KEY_INDEX@/${HL_SEC_KEY_INDEX}/g;";
+#ITS_SED_PARAMS+="s/@HL_SEC_FW_VERSION@/${HL_SEC_FW_VERSION}/g;";
+ITS_SED_PARAMS+="s/@SHA_VERSION@/${SHA_VERSION}/g;"
+ITS_SED_PARAMS+="s#@BINS_DIR_PATH@#${BINS_DIR}#g;"
+ITS_SED_PARAMS+="s#@BINS_PROD_DIR_PATH@#${BINS_PROD_DIR}#g;"
+ITS_SED_PARAMS+="s/@DEV_PREFIX@/${DEV_PREFIX}/g;"
diff --git a/board/habanalabs/jgs/post-image-fit.sh b/board/habanalabs/jgs/post-image-fit.sh
new file mode 100755
index 0000000..aae75a1
--- /dev/null
+++ b/board/habanalabs/jgs/post-image-fit.sh
@@ -0,0 +1,72 @@
+#!/usr/bin/env bash
+
+OBJCOPY=arc-linux-objcopy
+
+DEV_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
+source ${DEV_BASE}/jgs-image-utility.sh
+
+hl_create_section_variables
+
+MGMT_IO_ICCM_SIZE=$((ARC_MGMT_ICCM_SIZE * 1024))
+MGMT_IO_DCCM_SIZE=$((ARC_MGMT_DCCM_SIZE * 1024))
+PID_IO_ICCM_SIZE=$((ARC_PID_ICCM_SIZE * 1024))
+PID_IO_DCCM_SIZE=$((ARC_PID_DCCM_SIZE * 1024))
+
+function hl_copy_keys {
+	local FILEPATH=$1
+
+	cp ${KEY_BASE}/key/${KEY_NAME}.* ${FILEPATH}
+
+	return
+}
+
+function hl_copy_binaries {
+	local FILEPATH=$1
+
+	# copy binaries from other HL targets required for boot fit creation
+	# TEMPORARY: arcpid isn't ready yet, so comment-out the following lines. Otherwise, the script fails.
+	#cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-iccm.bin.le
+	#cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-dccm.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_NAME}/images/${DEV_PREFIX}-zephyr-mgmt-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_NAME}/images/${DEV_PREFIX}-zephyr-mgmt-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin.le
+	#cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-agent_fw_loader-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-agent_fw_loader-iccm.bin.le
+	#cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-agent_fw_loader-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-agent_fw_loader-dccm.bin.le
+
+	return
+}
+
+pushd ${STAGE_DIR}
+	hl_create_zephyr_bins_config
+	hl_extract_zephyr_elf_sections
+popd
+
+MGMT_STAT=${STATS_DIR}/zephyr_mgmt.stat
+# TEMPORARY: arcpid isn't ready yet, so comment-out the following line. Otherwise, the script fails.
+#PID_STAT=${STATS_DIR}/zephyr_pid.stat
+
+# Extract the start address from stat file
+ENTRY_ADDRESS=$(cat $MGMT_STAT | grep "Entry point address" | head -1)
+MGMT_STARTADDR=$(echo $ENTRY_ADDRESS | awk -F: '{ print $2 }' | xargs)
+# TEMPORARY: arcpid isn't ready yet, so comment-out the following line. Otherwise, the script fails.
+#ENTRY_ADDRESS=$(cat $PID_STAT | grep "Entry point address" | head -1)
+PID_STARTADDR=$(echo $ENTRY_ADDRESS | awk -F: '{ print $2 }' | xargs)
+
+# Start address is only 2 bytes. Perform OR operation with
+# ARC_PID_ICCM_LO to get 32 bit address and XX_STARTADDR_HI
+# is same as ARC_XX_ICCM_HI.
+PID_STARTADDR_LO=$((ARC_PID_ICCM_LO | PID_STARTADDR))
+MGMT_STARTADDR_LO=$((ARC_MGMT_ICCM_LO | MGMT_STARTADDR))
+
+ITS_SED_PARAMS+="s/@PID_STARTADDR_HI@/${ARC_PID_ICCM_HI}/g;"
+ITS_SED_PARAMS+="s/@PID_STARTADDR_LO@/${PID_STARTADDR_LO}/g;"
+ITS_SED_PARAMS+="s/@MGMT_STARTADDR_HI@/${ARC_MGMT_ICCM_HI}/g;"
+ITS_SED_PARAMS+="s/@MGMT_STARTADDR_LO@/${MGMT_STARTADDR_LO}/g;"
+ITS_SED_PARAMS+="s/@MGMT_IO_ICCM_SIZE@/${MGMT_IO_ICCM_SIZE}/g;"
+ITS_SED_PARAMS+="s/@MGMT_IO_DCCM_SIZE@/${MGMT_IO_DCCM_SIZE}/g;"
+ITS_SED_PARAMS+="s/@PID_IO_ICCM_SIZE@/${PID_IO_ICCM_SIZE}/g;"
+ITS_SED_PARAMS+="s/@PID_IO_DCCM_SIZE@/${PID_IO_DCCM_SIZE}/g;"
+ITS_SED_PARAMS+="s/@MGMT_IO_BASEADDR_HI@/${ARC_MGMT_CPU_HI}/g;"
+ITS_SED_PARAMS+="s/@MGMT_IO_BASEADDR_LO@/${ARC_MGMT_CPU_LO}/g;"
+ITS_SED_PARAMS+="s/@PID_IO_BASEADDR_HI@/${ARC_PID_CPU_HI}/g;"
+ITS_SED_PARAMS+="s/@PID_IO_BASEADDR_LO@/${ARC_PID_CPU_LO}/g;"
+
diff --git a/board/habanalabs/jgs/post-image-spi.sh b/board/habanalabs/jgs/post-image-spi.sh
new file mode 100755
index 0000000..afc0669
--- /dev/null
+++ b/board/habanalabs/jgs/post-image-spi.sh
@@ -0,0 +1,747 @@
+#!/usr/bin/env bash
+
+OPENSSL=/usr/bin/openssl
+OBJCOPY=arc-linux-objcopy
+SHA_CHKSUM=sha384sum
+
+FAKE_FUSE_COMPUTE_D0_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/compute_d0_fake_fuse.txt
+FAKE_FUSE_COMPUTE_D1_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/compute_d1_fake_fuse.txt
+FAKE_FUSE_COMPUTE_D2_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/compute_d2_fake_fuse.txt
+FAKE_FUSE_COMPUTE_D3_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/compute_d3_fake_fuse.txt
+FAKE_FUSE_IO_D0_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/io_d0_fake_fuse.txt
+FAKE_FUSE_IO_D1_TEXT_FILE=${BR2_EXTERNAL}/board/habanalabs/jgs/io_d1_fake_fuse.txt
+
+SPEC_BOOT_DESC_HDR_FILE=${BR2_EXTERNAL}/../specs/fs/jgs/jgs_boot_info.h
+SPEC_SECURITY_HDR_FILE=${BR2_EXTERNAL}/../specs/fs/jgs/jgs_security.h
+SPEC_HL_BOOT_DESC_HDR_FILE=${BR2_EXTERNAL}/../specs/fs/common/hl_boot_desc.h
+
+PREBOOT_PADDED_BINFILE=${PREBOOT_PADDED_BINFILE:-preboot.bin.pad}
+
+SECONDARY_MANIFEST_OFFSET=0x200000
+
+PPBOOT_PRI_OFFSET=0x10000
+PPBOOT_SEC_OFFSET=$((PPBOOT_PRI_OFFSET + SECONDARY_MANIFEST_OFFSET))
+PPBOOT_TEXT_BASE=0x300007FF89A0000
+PPBOOT_START_ADDR=0x300007FF89A0000
+PPBOOT_ALT_DMA_ADDR=0x300007FF8A08000
+
+PREBOOT_PRI_OFFSET=0x30000
+PREBOOT_SEC_OFFSET=$((PREBOOT_PRI_OFFSET + SECONDARY_MANIFEST_OFFSET))
+PREBOOT_TEXT_BASE=0x300007FF8A08000
+PREBOOT_START_ADDR=0x300007FF8A08000
+PREBOOT_CPU_BASE_ADDR=0x79CE0000	# mmIO_0_CFG_HALF_0_PSOC_ARC_MNG_ARC_BASE
+PREBOOT_IMAGE_FLAGS=0x7E		# load and exec for all rst types
+
+FAKE_FUSE_PRI_OFFSET=0x140000
+FAKE_FUSE_SEC_OFFSET=$((FAKE_FUSE_PRI_OFFSET + SECONDARY_MANIFEST_OFFSET))
+FAKE_FUSE_TEXT_BASE=0x0
+FAKE_FUSE_START_ADDR=0x0
+FAKE_FUSE_BINFILE=${DEV_PREFIX}-fake-fuse.bin.le
+
+# refer fw_boot_pointer_type
+IMAGE_TYPE_SPI=0
+IMAGE_TYPE_FAKE_FUSE=1
+IMAGE_TYPE_EROM=2	# TODO: Need to remove EROM related Implementation.
+
+SECURITY_VER_FILE=${SPECS_EMBEDDED_ROOT}/fs/${DEV_PREFIX}/${DEV_PREFIX}_security.h
+NUM_PROD_KEYS=0
+
+ARCBOOT_BINFILE_ORIG=${ARCBOOT_BINFILE_PAD:-arcboot_rom-orig.bin}
+
+function hl_copy_binaries {
+	# TODO fill with content
+
+	# creating temporary dummy files
+	touch jgs-erom.bin.le
+	touch jgs-images-pointers.bin.le
+	head -c 10000 /dev/urandom > jgs-ppboot-pcie_phy_fw.bin.le.int
+
+	cp ${KEY_BASE}/key/hl_${KEY_BASE_NAME}.* .
+
+	return
+}
+
+function copy_all_keys {
+	cp ${KEY_BASE}/key/*.dtb ${1:-.}
+	cp ${KEY_BASE}/key/*.key ${1:-.}
+	cp ${KEY_BASE}/key/*.crt ${1:-.}
+}
+
+function generate_public_keys {
+	# TODO fill with content
+	return
+}
+
+function prep_image_pcie_pif {
+	# TODO fill with content
+	return
+}
+
+function prep_image_erom {
+	# TODO fill with content
+	return
+}
+
+function export_desc_var {
+	FAKE_FUSE_DIE_PARAM="${1}${2}"
+	export "$FAKE_FUSE_DIE_PARAM"=${3}
+}
+
+function get_aligned_size {
+	local param=$1
+	local alignment=1024
+
+	if [[ "$param" =~ ^[0-9]+$ ]]; then
+		local size=$param
+	elif [ -f "$param" ]; then
+		local size=$(stat -Lc "%s" $(eval "echo \$param"))
+	else
+		echo "Error: Unexpected entry for size alignment"
+		exit 1
+	fi
+
+	local remainder=$((size % $alignment))
+	if [ $remainder -eq 0 ]; then
+		echo $size
+	else
+		echo $(($size + $alignment - $remainder))
+	fi
+}
+
+function prep_image_fake_fuse {
+	local OUTF=$1
+	local FAKE_FUSE_TMP_BIN=die_fake_fuse.bin
+
+	get_fake_fuse_bin_list_array ${FW_IF_HEADER} "fake_fuse_bin_index"
+
+	# Boot pointer struct size + (Image count * pointer size)
+	local BOOT_POINTER_SIZE_ALLOC=$((IMG_PTR_SZ + (${#FAKE_FUSE_ARRAY[@]} * 4)))
+	# Boot pointer struct size + (Image count * pointer size) + (Img desc * Image count)
+	BOOT_POINTER_SIZE_ALLOC=$((BOOT_POINTER_SIZE_ALLOC + (IMG_DESC_SZ * ${#FAKE_FUSE_ARRAY[@]})))
+	local align_sz=16
+	#ALIGN(payload)
+	BOOT_POINTER_SIZE_ALLOC=$(((BOOT_POINTER_SIZE_ALLOC+align_sz-1)/align_sz*align_sz))
+	#ALIGN(payload) + CRC32
+	BOOT_POINTER_SIZE_ALLOC=$((BOOT_POINTER_SIZE_ALLOC+4))
+
+	# Binary insertion alignment
+	BOOT_POINTER_SIZE_ALLOC="$(get_aligned_size $BOOT_POINTER_SIZE_ALLOC)"
+
+	local CUR_OFFSET_IN_BIN=$BOOT_POINTER_SIZE_ALLOC
+	local FUSE_BIN_SIZE=0
+
+	# Create 16Byte aligned 6 Fuse binaries from text files.
+	for FUSE_FILE_INDEX in "${!FAKE_FUSE_ARRAY[@]}"; do
+		FAKE_FUSE_DIE="${FAKE_FUSE_ARRAY[$FUSE_FILE_INDEX]}"
+
+		FAKE_FUSE_DIE_TEXT_FILE_VAR="${FAKE_FUSE_DIE}_TEXT_FILE" # Storing variable name
+
+		FAKE_FUSE_DIE_TEXT_FILE_NAME=$(basename "${!FAKE_FUSE_DIE_TEXT_FILE_VAR}")
+		FAKE_FUSE_DIE_BINFILE_NAME="${FAKE_FUSE_DIE_TEXT_FILE_NAME%.txt}.bin"
+
+		#Bank 0
+		xxd -r -p ${!FAKE_FUSE_DIE_TEXT_FILE_VAR} > ${FAKE_FUSE_TMP_BIN}
+		#Bank 1
+		xxd -r -p ${!FAKE_FUSE_DIE_TEXT_FILE_VAR} >> ${FAKE_FUSE_TMP_BIN}
+		# Prepare Fuse binary with 2 Banks
+		dd status=none conv=sync bs=16 if=${FAKE_FUSE_TMP_BIN}	> ${FAKE_FUSE_DIE_BINFILE_NAME}
+
+		# Insert Die Fake fuse bin file.
+		FUSE_BIN_SIZE="$(get_aligned_size $FAKE_FUSE_DIE_BINFILE_NAME)"
+		dd status=none conv=sync seek=$(printf %d ${CUR_OFFSET_IN_BIN}) if=$FAKE_FUSE_DIE_BINFILE_NAME bs=1 of=${OUTF}
+
+		export_desc_var $FAKE_FUSE_DIE "_BINFILE" $FAKE_FUSE_DIE_BINFILE_NAME
+		export_desc_var $FAKE_FUSE_DIE "_TEXT_BASE" 0x0
+		export_desc_var $FAKE_FUSE_DIE "_START_ADDR" 0x0
+		export_desc_var $FAKE_FUSE_DIE "_PRI_OFFSET" $CUR_OFFSET_IN_BIN
+		export_desc_var $FAKE_FUSE_DIE "_SEC_OFFSET" $((CUR_OFFSET_IN_BIN + SECONDARY_MANIFEST_OFFSET))
+
+		CUR_OFFSET_IN_BIN=$((CUR_OFFSET_IN_BIN + FUSE_BIN_SIZE))
+	done
+	rm -f $FAKE_FUSE_TMP_BIN
+
+	# Create fake fuse boot pointer bin.
+	prep_image_ptrs_common_v1 ${DEV_PREFIX}-fake-fuse-pointers.bin.le $IMAGE_TYPE_FAKE_FUSE
+	FILE_SIZE=$(get_aligned_size "${DEV_PREFIX}-fake-fuse-pointers.bin.le")
+
+	# If Boot pointer bin size greater than BOOT_POINTER_SIZE_ALLOC, break the execution.
+	if [ "$FILE_SIZE" -gt "$BOOT_POINTER_SIZE_ALLOC" ]; then
+		echo "Error: Fake fuse Boot pointer construction failed"
+		exit 1
+	fi
+
+	# Insert Fake fuse pointer bin at FAKE_FUSE starting.
+	dd status=none conv=notrunc seek=0 if=${DEV_PREFIX}-fake-fuse-pointers.bin.le bs=1 of=${OUTF}
+}
+
+function zephyr_preboot_pack {
+	local PREBOOT_BINFILE=$1
+	local PRIME_IMAGE_RAM_KEY=$2
+	local PRIME_IMAGE_RAM_CRC_ERROR=$3
+	local PRIME_IMAGE_RAM_SIGN_ERROR=$4
+
+	hl_pack_bin ${PREBOOT_BINFILE} 64K ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+
+	return
+}
+
+function hl_join_bin {
+	NAME_1=$1
+	NAME_2=$2
+	PACK_NAME=$3
+	# make file lineup to 16byte chunks
+	dd status=none conv=sync bs=16 if=${NAME_1}.bin		> ${NAME_1}.bin.algn
+	cat ${NAME_1}.bin.algn ${NAME_2}.bin	> ${PACK_NAME}
+	rm ${NAME_1}.bin.* -rf
+}
+
+# TODO: JGS
+MAX_SPI_DMA_DATA_TRANSFER=65520
+SPI_DMA_DEST_64KB_OFFS=65536
+function hl_pack_bin {
+	FILEFULLPATH=$1
+	FILENAME=$(basename ${FILEFULLPATH})
+	FILEPATH=$(dirname ${FILEFULLPATH})
+	BSIZE=$2
+	KEY_NUM=$3
+	CRC_ERROR=$4
+	SIGN_ERROR=$5
+
+	PADFULLBINFILE=${FILEFULLPATH}.pad
+	dd status=none conv=sync bs=16 if=${FILEFULLPATH} of=${PADFULLBINFILE}
+
+	SIZE_WITHOUT_CRC=$(stat --printf="%s" ${PADFULLBINFILE})
+	NUM_SLICES=$((SIZE_WITHOUT_CRC/MAX_SPI_DMA_DATA_TRANSFER))
+	MODULO_SLICES=$((SIZE_WITHOUT_CRC%MAX_SPI_DMA_DATA_TRANSFER))
+	if [ ${MODULO_SLICES} -ne 0 ]; then
+		NUM_SLICES=$((NUM_SLICES+1))
+	fi
+
+	# image size to be computed as part of image descriptor function
+	local leftover=${SIZE_WITHOUT_CRC}
+	local skip_offset=0
+
+
+	# image packing example:
+	# image_size = 200kb + 16 bytes CRC = 0x32010
+	# num_slice: 4
+	# 1st slice: flash offset: 0x0 DST: 0x0 (dma_size: 0xFFF4 left_size: 0x2201C)
+	# 2nd slice: flash offset: 0x10000 DST: 0xFFF0 (dma_size: 0xFFF4 left_size: 0x12028)
+	# 3nd slice: flash offset: 0x20000 DST: 0x1FFE0 (dma_size: 0xFFF4 left_size: 0x2034)
+	# 4th slice: flash offset: 0x30000 DST: 0x2FFD0 (dma_size: 0x2034 left_size: 0x0)
+	#
+	#  =========================================================
+	# | PPBOOT IMG SLICE1 + CRC + PADDING (0xFFF0 + 0x4 + 0x12) |
+	#  ---------------------------------------------------------
+	# | PPBOOT IMG SLICE2 + CRC + PADDING (0xFFF0 + 0x4 + 0x12) |
+	#  ---------------------------------------------------------
+	# | PPBOOT IMG SLICE3 + CRC + PADDING (0xFFF0 + 0x4 + 0x12) |
+	#  ---------------------------------------------------------
+	# | PPBOOT IMG SLICE4 + CRC (0x2030 + 0x4)		    |
+	#  =========================================================
+	#
+	# Note: Padding size is not included in total size as DMA is
+	# not expected to copy padding data.
+	for ((slice_cnt=0; slice_cnt<${NUM_SLICES}; slice_cnt++)); do
+		# bin file: (64KB-16) + CRC
+		local BIN_FILE=${PADFULLBINFILE}.part${slice_cnt}
+		if [ $leftover -lt $MAX_SPI_DMA_DATA_TRANSFER ]; then
+			size_bytes=$leftover
+		else
+			size_bytes=$MAX_SPI_DMA_DATA_TRANSFER
+		fi
+
+		size_words=$((size_bytes/4))
+		dd status=none skip=${skip_offset} conv=sync bs=4 if=${PADFULLBINFILE} count=${size_words} > ${BIN_FILE}
+		hl_calc_crc32 ${BIN_FILE}
+		if [ ${CRC_ERROR} -eq 0 ]; then
+			cat ${BIN_FILE}.crc32 >> ${BIN_FILE}
+		fi
+		flash_offset=$((SPI_DMA_DEST_64KB_OFFS/4))
+		flash_offset=$((flash_offset*slice_cnt))
+		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
+		leftover=$((leftover-size_bytes))
+		skip_offset=$((skip_offset+size_words))
+	done
+
+	hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int ${FILEPATH}/${DEV_PREFIX}-${FILENAME}.le 4
+	rm -f ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
+}
+
+function hl_sign_bin {
+	BASENAME=$1
+	CERT=$2
+	$OPENSSL dgst -sha384 -sign ${CERT}.key -out ${BASENAME}.sig ${BASENAME}
+}
+
+function extract_key_props {
+	KEY_NAME=$1
+	PROP_NAME=$2
+	OUT_NAME=$3
+
+	TMPNAME=${STAGE_DIR}/${KEY_NAME}-${PROP_NAME}.txt
+	fdtget ${BINS_PROD_DIR}/${KEY_NAME}.dtb /signature/key-${KEY_NAME} rsa,${PROP_NAME} -t bx > ${TMPNAME}
+	sed 's/\s\+/\n/g' ${TMPNAME} > ${TMPNAME}.tmp
+	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${TMPNAME}.tmp  > ${TMPNAME}.tmp.le
+	xxd -r -p ${TMPNAME}.tmp.le > ${OUT_NAME}
+	rm -f ${TMPNAME}.*
+}
+
+function hl_get_key_props {
+	KEY_NUM=$1
+
+	if [ ${KEY_NUM} -eq 0 ]; then
+		KEY_NAME=$(echo -n "hl_${KEY_BASE_NAME}")
+	else
+		KEY_NAME=$(echo -n "cust${KEY_NUM}_${KEY_BASE_NAME}")
+	fi
+
+	extract_key_props ${KEY_NAME} r-squared ${STAGE_DIR}/rr.prop
+
+	extract_key_props ${KEY_NAME} modulus ${STAGE_DIR}/mod.prop
+
+	extract_key_props ${KEY_NAME} exponent ${STAGE_DIR}/exp.prop
+
+	extract_key_props ${KEY_NAME} n0-inverse ${STAGE_DIR}/inv.proptemp
+	swap_file ${STAGE_DIR}/inv.proptemp ${STAGE_DIR}/inv.prop
+
+	extract_key_props ${KEY_NAME} num-bits ${STAGE_DIR}/nb.proptemp
+	swap_file ${STAGE_DIR}/nb.proptemp ${STAGE_DIR}/nb.prop
+
+	#no separate public key will be part of the image, since these props serve the purpose.
+	# Hash will be computed on the Props itself
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop	> ${BINS_PROD_DIR}/${KEY_NAME}.pub
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop	> ${BINS_PROD_DIR}/${KEY_NAME}-key-props.bin
+
+	rm -f ${STAGE_DIR}/*.txt ${STAGE_DIR}/*.proptemp ${STAGE_DIR}/*.prop
+}
+
+function hl_chksum_padding {
+	# Add zero padding for compatibility with image descriptors size.
+	local pad_dwords=$1
+	local pad_data=0
+
+	for i in $(seq 1 $pad_dwords); do
+		printf "%08x" $pad_data
+	done
+}
+
+function prep_img_desc_with_crc {
+	BINFILE=$1
+	OUTFILENAME=$(basename $1)
+	cat ${BINFILE}>${BINFILE}.tmp.pad
+	dd status=none conv=sync bs=16 if=${BINFILE}.tmp.pad of=${BINFILE}.tmp.algn
+	hl_calc_crc32 ${BINFILE}.tmp.algn
+	cat ${BINFILE}.tmp.algn ${BINFILE}.tmp.algn.crc32 > ${BINS_DIR}/${DEV_PREFIX}-${OUTFILENAME}
+	rm -rf ${BINFILE}.tmp*
+}
+
+function hl_chksum_add {
+	local checksum=$1
+
+	# Update the sha384 checksum field
+	echo $checksum
+}
+
+function print_image_chksum {
+	local name=${1^^}
+	local BINFILE=$(eval "echo \${${name}_BINFILE}")
+
+	#align the source binaries to 16 bytes
+	dd status=none conv=sync bs=16 if=${BINFILE} of=${BINFILE}.tmp.pad
+	# Prepare checksum
+	local CKSUM=`sha384sum ${BINFILE}.tmp.pad`
+	hl_chksum_add ${CKSUM}
+	rm -f ${BINFILE}.tmp.pad
+}
+
+function prep_img_crc32 {
+	local name=${1^^}
+	local BINFILE=$(eval "echo \${${name}_BINFILE}")
+
+	# reverse CRC32 source as DMA engine works like that
+	hl_reverse_words ${BINFILE} ${BINFILE}.out 4
+	CRC32=$(crc32 ${BINFILE}.out)
+	echo ${CRC32}
+}
+
+function get_spi_component_ver() {
+	local img_name=${1^^}
+	local baremetal_comp=("PPBOOT" "EROM_PCIE" "EROM_RECOVERY")
+	local preboot_comp=("PREBOOT")
+	local ver
+
+	if [ $img_name == $preboot_comp ]; then
+		ver=$(cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}/build/preboot-custom/localversion-habanalabs)
+		echo ${ver#*-}
+		return
+	fi
+
+	for i in "${baremetal_comp[@]}"; do
+		if [ $img_name == $i ]; then
+			ver=$(cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}/build/baremetal-custom/localversion-habanalabs)
+			echo ${ver#*-}
+			return
+		fi
+	done
+
+	echo "no version"
+}
+
+function print_image_version {
+	local img_name=${1}
+	local img_ver_str=$(get_spi_component_ver $img_name)
+	local max_ver_len=128
+
+	echo -n $img_ver_str | xxd -p
+
+	#pad the remaining space with zeroes
+	local ver_len=${#img_ver_str}
+	local ver_pad=$(($max_ver_len - $ver_len))
+	if [ $ver_len -eq $max_ver_len ]; then
+		return;
+	fi
+
+	for i in $(seq 1 $ver_pad); do
+		printf "%02x" 0
+	done
+}
+
+function print_image_desc {
+	local img_name=$1
+	local img_num=$2
+	local -n spi_reg_offsets=$3
+	local img_base_pri=$(get_image_base "$img_name"_PRI)
+	local img_base_sec=$(get_image_base "$img_name"_SEC)
+	local img_load=$(get_image_loadaddr $img_name)
+	local img_loadh=$((img_load >> 32))		# take upper bytes
+	local img_loadl=$((img_load & 0xFFFFFFFF))	# take lower bytes
+	local img_dma_alt_h_0=${img_loadh}
+	local img_dma_alt_l_0=${img_loadl}
+	local img_dma_alt_h_1=${img_loadh}
+	local img_dma_alt_l_1=${img_loadl}
+	local img_name_2=$img_name
+	local img_flags=0
+	local img_cpu_base=0
+	local img_mem_base_l=0
+	local img_mem_base_h=0
+	local img_mem_size=0
+
+	if [ $img_name == "PPBOOT" ]
+	then
+		img_name_2="PPBOOT_PCIE_PHY_PADDED"
+		# prefetch location address for secondary ppboot image
+		img_dma_alt_h_1=$((PPBOOT_ALT_DMA_ADDR >> 32))		# take upper bytes
+		img_dma_alt_l_1=$((PPBOOT_ALT_DMA_ADDR & 0xFFFFFFFF))	# take lower bytes
+		img_base_pri=${spi_reg_offsets[PPBOOT_PRIMARY]}
+		img_base_sec=${spi_reg_offsets[PPBOOT_SECONDARY]}
+	fi
+
+	if [ $img_name == "PREBOOT" ]
+	then
+		img_base_pri=${spi_reg_offsets[PREBOOT_IO_DIE_PRIMARY]}
+		img_base_sec=${spi_reg_offsets[PREBOOT_IO_DIE_SECONDARY]}
+		img_cpu_base=${PREBOOT_CPU_BASE_ADDR}
+		img_flags=${PREBOOT_IMAGE_FLAGS}
+	fi
+
+	if [ $img_name == "FAKE_FUSE" ]
+	then
+		img_base_pri=${spi_reg_offsets[FAKE_FUZE_PRIMARY]}
+		img_base_sec=${spi_reg_offsets[FAKE_FUZE_SECONDARY]}
+	fi
+
+	if [ $img_name == "EROM_PCIE" -o $img_name == "EROM_RECOVERY" -o $img_name == "PREBOOT" -o $img_name == "PREBOOT_D1" ]
+	then
+		img_name_2=${img_name}_PADDED
+	fi
+
+	local img_size=$(get_image_size $img_name_2)
+	local start_addr=$(get_image_startaddr $img_name)
+	local start_addr_h=$((start_addr >> 32))		# take upper bytes
+	local start_addr_l=$((start_addr & 0xFFFFFFFF))	# take lower bytes
+	local size_corr_fac=0
+
+	if [[ ${ipnames[$i]} != "FAKE_FUSE"* ]];
+	then
+		local num_slices=$((img_size / MAX_SPI_DMA_DATA_TRANSFER))
+		local modulo_slices=$((img_size % MAX_SPI_DMA_DATA_TRANSFER))
+		if [ ${modulo_slices} -ne 0 ]; then
+			((++num_slices))
+		fi
+
+		size_corr_fac=$((num_slices * 4))
+	fi
+
+	local img_size_final=$((img_size+size_corr_fac))
+
+	# Prepare descriptor magic "PTR"0x$img_num
+	printf "%x%x%x%02x" "'P" "'T" "'R" $img_num | rev_bytes_stdin
+
+	# Prepare descriptor version
+	local ver=$(grep CUR_VER ${SPEC_BOOT_DESC_HDR_FILE} | awk '{print $3}')
+	printf "%08x" $ver | rev_bytes_stdin
+
+	# Array of spi offsets, for primary and secondary images
+	printf "%08x" $img_base_pri | rev_bytes_stdin
+	printf "%08x" $img_base_sec | rev_bytes_stdin
+
+	# Prepare image size
+	printf "%08x" $img_size_final | rev_bytes_stdin
+
+	# If one of the addresses is empty use 0x0 instead
+	# Prepare image load address high
+	printf "%08x" $img_loadh | rev_bytes_stdin
+	# Prepare image load address low
+	printf "%08x" $img_loadl | rev_bytes_stdin
+
+	prep_img_crc32 $img_name_2
+
+	print_image_chksum $img_name_2
+
+	# Prepare image start address high
+	printf "%08x" $start_addr_h | rev_bytes_stdin
+	# Prepare image start address low
+	printf "%08x" $start_addr_l | rev_bytes_stdin
+
+	# Prepare default image dma address high
+	printf "%08x" $img_dma_alt_h_0 | rev_bytes_stdin
+	# Prepare alternative image dma address high
+	printf "%08x" $img_dma_alt_h_1 | rev_bytes_stdin
+	# Prepare default image dma address low
+	printf "%08x" $img_dma_alt_l_0 | rev_bytes_stdin
+	# Prepare alternative image dma address low
+	printf "%08x" $img_dma_alt_l_1 | rev_bytes_stdin
+
+	# Prepare component version
+	print_image_version $img_name
+
+	# Prepare autoload parameters
+	printf "%08x" $img_cpu_base | rev_bytes_stdin
+	printf "%08x" $img_mem_base_h | rev_bytes_stdin
+	printf "%08x" $img_mem_base_l | rev_bytes_stdin
+	printf "%08x" $img_mem_size | rev_bytes_stdin
+	printf "%08x" $img_flags | rev_bytes_stdin
+}
+
+function prep_image_ptrs_common_v1 {
+	local img_file=$1
+	local out_file=$1
+	local KEY_NUM=0
+	local IMG_TYPE=$2
+	declare -A SPI_REGION_OFFSETS
+	declare -A SPI_REGION_SIZES
+
+	if [ ${IMG_TYPE} == ${IMAGE_TYPE_FAKE_FUSE} ]
+	then
+		KEY_NUM=0 # 255 to stop, < 255 to write slave Fake fuse values.
+	fi
+
+	hl_parse_spi_flash_regions SPI_REGION_OFFSETS SPI_REGION_SIZES jgs
+
+	# Prepare empty space for size of boot pointer and size of image pointer.
+	echo -ne '\x00\x00\x00\x00\x00\x00\x00\x00' > $img_file
+
+	local ipnames=0
+
+	# The elements in the array should be in the order of hl_fw_img_type enums
+	local ipnames_spi_img=("PREBOOT" "PPBOOT" "FAKE_FUSE")
+	local ipnames_erom=("EROM_PCIE" "EROM_RECOVERY")
+
+	# Image pointers
+	if [ ${IMG_TYPE} == ${IMAGE_TYPE_EROM} ]
+	then
+		ipnames=("${ipnames_erom[@]}")
+		# Pointers structure magic "NXT"0x0
+		printf "%x%x%x00" "'N" "'X" "'T" | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
+	elif [ ${IMG_TYPE} == ${IMAGE_TYPE_SPI} ]
+	then
+		ipnames=("${ipnames_spi_img[@]}")
+		# Pointers structure magic "NXT"0x0
+		printf "%x%x%x00" "'N" "'X" "'T" | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
+	elif [ ${IMG_TYPE} == ${IMAGE_TYPE_FAKE_FUSE} ]
+	then
+		ipnames=("${FAKE_FUSE_ARRAY[@]}")
+		# Fake Fuse Low Voltage Attack magic "LVA"0x2
+		printf "%x%x%x%02x" "'L" "'V" "'A" $IMG_TYPE | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
+	else
+		echo "Error: wrong Image type"
+		exit 1
+	fi
+
+	local ipnum=${#ipnames[*]}
+
+	local offs_corr_fac=0
+
+	printf "%08x" $ipnum | rev_bytes_stdin | xxd -r -p -s +12 - $img_file
+
+	local pub_key_size=784
+	#Prepare pub key size
+	printf "%08x" $pub_key_size | rev_bytes_stdin | xxd -r -p -s +16 - $img_file
+
+	local sig_size=384
+	#Prepare signature size
+	printf "%08x" $sig_size | rev_bytes_stdin | xxd -r -p -s +20 - $img_file
+
+	local sec_version=$(grep FW_VERSION ${SPEC_SECURITY_HDR_FILE} | awk '{print $3}')
+	#Prepare security version
+	printf "%02x" $sec_version | rev_bytes_stdin | xxd -r -p -s +24 - $img_file
+
+	local key_idx=$KEY_NUM
+	#Prepare key index
+	printf "%02x" $key_idx | rev_bytes_stdin | xxd -r -p -s +25 - $img_file
+
+	local boot_ptrs_version=$(grep HL_FW_BOOT_VERSION ${SPEC_HL_BOOT_DESC_HDR_FILE} | awk '{print $3}')
+	#Prepare boot_ptrs_version
+	printf "%02x" $boot_ptrs_version | rev_bytes_stdin | xxd -r -p -s +26 - $img_file
+
+	local boot_ptrs_size=$IMG_PTR_SZ
+	#Prepare boot_ptrs_size
+	printf "%02x" $boot_ptrs_size | rev_bytes_stdin | xxd -r -p -s +27 - $img_file
+
+	local offset_array=$IMG_PTR_SZ
+	#Prepare security version for Stable image
+	printf "%08x" $offset_array | rev_bytes_stdin | xxd -r -p -s +28 - $img_file
+
+	# Pointer structure offsets
+	local img_descs_base_off=$((IMG_PTR_SZ + (ipnum * 4)))
+	for ((i=0; i<$ipnum; i++)); do
+		local offset=$((img_descs_base_off + (i * IMG_DESC_SZ)))
+		local offloc=$((IMG_PTR_SZ + (i * 4)))
+
+		# Prepare pointer (offset) to descriptor
+		printf "%08x" $offset | rev_bytes_stdin | xxd -r -p -s +$offloc - $img_file
+
+		# Prepare descriptor
+		print_image_desc ${ipnames[$i]} $((i + 1)) SPI_REGION_OFFSETS | xxd -r -p -s +$offset - $img_file
+	done
+
+	if [ ${IMG_TYPE} != ${IMAGE_TYPE_FAKE_FUSE} ]
+	then
+		if [ ${KEY_NUM} -eq 0 ]; then
+			KEY_NAME=$(echo -n "hl_${KEY_BASE_NAME}")
+		else
+			KEY_NAME=$(echo -n "cust${KEY_NUM}_${KEY_BASE_NAME}")
+		fi
+
+		#Extract Key Properties and store them in a bin file
+		hl_get_key_props ${KEY_NUM}
+
+		TEMPNAME=${FUNCNAME}
+		TEMPFULLNAME=${STAGE_DIR}/${FUNCNAME}
+		dd status=none conv=sync bs=16 if=$img_file			> ${TEMPFULLNAME}
+
+		PKFILE=${TEMPNAME}.payload.key
+		PKFULLFILE=${TEMPFULLNAME}.payload.key
+
+		cat ${TEMPFULLNAME} ${BINS_PROD_DIR}/${KEY_NAME}-key-props.bin > ${PKFULLFILE}
+
+		#The size of the complete image(Payload + Key + Signature) needs to be calculated
+		#and placed in the manifest even before hashing and signing over it as the size
+		#data itself needs to be hashed. Due to this we are required to calculate the
+		#complete image size to be dmaed by considering payload + key + signature + CRC size.
+		#PL NOTE: We also need to consider the 16 byte alignment in case the resultant
+		#size of (payload + key + signature) is not 16 byte aligned. This is because
+		#of dma requirement to place the crc at the end of a 16 byte aligned image.
+		#Hence the final formula = ALIGN_16(Payload + Key + Signature) + CRC
+
+		#payload + key size
+		local pk_sz=$(stat -Lc "%s" $(eval "echo \${PKFULLFILE}"))
+		#size to be aligned to
+		local align_sz=16
+		#payload + key + signature size
+		local pk_sig_sz=$((pk_sz+sig_size))
+		#ALIGN(payload + key + signature size)
+		local pk_sig_align_sz=$(((pk_sig_sz+align_sz-1)/align_sz*align_sz))
+		#ALIGN(payload + key + signature size) + CRC size
+		printf "%08x" $((pk_sig_align_sz+4)) | rev_bytes_stdin | xxd -r -p -s +0 - ${PKFILE}
+
+		printf "%08x" $IMG_DESC_SZ | rev_bytes_stdin | xxd -r -p -s +4 - ${PKFILE}
+
+		$OPENSSL dgst -sha384 -binary ${PKFULLFILE} > ${PKFULLFILE}.sha384
+
+		hl_sign_bin ${PKFULLFILE} ${STAGE_DIR}/${KEY_NAME}
+
+		SIGBINFILE=${PKFILE}.sig.bin
+
+		cat ${PKFULLFILE} ${PKFULLFILE}.sig > ${STAGE_DIR}/${SIGBINFILE}
+
+		prep_img_desc_with_crc ${STAGE_DIR}/${SIGBINFILE}
+
+		hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-${SIGBINFILE} $out_file 4
+	else
+		local boot_ptr_size=$(stat -Lc "%s" $(eval "echo \${img_file}"))
+		local align_sz=16
+		#ALIGN(payload)
+		boot_ptr_size=$(((boot_ptr_size+align_sz-1)/align_sz*align_sz))
+		#ALIGN(payload) + CRC size
+		printf "%08x" $((boot_ptr_size+4)) | rev_bytes_stdin | xxd -r -p -s +0 - ${img_file}
+		printf "%08x" $IMG_DESC_SZ | rev_bytes_stdin | xxd -r -p -s +4 - ${img_file}
+		prep_img_desc_with_crc ${img_file}
+		hl_reverse_words ${DEV_PREFIX}-${img_file} $out_file 4
+	fi
+}
+
+function prep_image_ptrs {
+	prep_image_ptrs_common_v1 $1 $IMAGE_TYPE_SPI
+	return
+}
+
+function calculate_hash {
+	# TODO fill with content
+
+	# creating temporary dummy hash result
+	touch ${BINS_PROD_DIR}/hl_rsa3072.pub.hash
+
+	return
+}
+
+function hl_pad_arcboot_rom {
+	cp ${BINS_DIR}/${ARCBOOT_BINFILE} ${BINS_DIR}/${ARCBOOT_BINFILE_ORIG}
+
+	# TODO: pad size different for compute
+	# pad arcboot rom binary to size of ROM memory
+	dd status=none conv=sync bs=128K if=${ARCBOOT_BINFILE_ORIG} > ${ARCBOOT_BINFILE}
+}
+
+# ToDo remove after agent fw check patches related to FSW-3432
+function create_spi_image {
+	local OUTF=${1}
+	local PPBOOT_BIN=${2}
+	local PREBOOT_BIN=${3}
+	local IMG_PTR_BIN=${4}
+	local FAKE_FUSE_BIN=${5}
+
+	declare -rA REGION_TO_FILENAME=(["IMAGE_DESC_PRIMARY"]=$IMG_PTR_BIN \
+					["PPBOOT_PRIMARY"]=$PPBOOT_BIN \
+					["PREBOOT_IO_DIE_PRIMARY"]=$PREBOOT_BIN \
+					["PREBOOT_COMPUTE_DIE_PRIMARY"]="FF" \
+					["CXL_ARC_ICCM_PRIMARY"]="FF" \
+					["CXL_ARC_DCCM_PRIMARY"]="FF" \
+					["DSM_ARC_ICCM_PRIMARY"]="FF" \
+					["DSM_ARC_DCCM_PRIMARY"]="FF" \
+					["RECOVERY_IMAGE_PCIE_PRIMARY"]="FF" \
+					["RECOVERY_IMAGE_FW_UPD_PRIMARY"]="FF" \
+					["FAKE_FUZE_PRIMARY"]=$FAKE_FUSE_BIN \
+					["RESERVED_1"]="FF" \
+					["IMAGE_DESC_SECONDARY"]=$IMG_PTR_BIN \
+					["PPBOOT_SECONDARY"]=$PPBOOT_BIN \
+					["PREBOOT_IO_DIE_SECONDARY"]=$PREBOOT_BIN \
+					["PREBOOT_COMPUTE_DIE_SECONDARY"]="FF" \
+					["CXL_ARC_ICCM_SECONDARY"]="FF" \
+					["CXL_ARC_DCCM_SECONDARY"]="FF" \
+					["DSM_ARC_ICCM_SECONDARY"]="FF" \
+					["DSM_ARC_DCCM_SECONDARY"]="FF" \
+					["RECOVERY_IMAGE_PCIE_SECONDARY"]="FF" \
+					["RECOVERY_IMAGE_FW_UPD_SECONDARY"]="FF" \
+					["FAKE_FUZE_SECONDARY"]=$FAKE_FUSE_BIN \
+					["RESERVED_2"]="FF")
+	declare -A SPI_REGION_OFFSETS
+	declare -A SPI_REGION_SIZES
+
+	hl_parse_spi_flash_regions SPI_REGION_OFFSETS SPI_REGION_SIZES fs1
+	hl_validate_spi_flash_layout REGION_TO_FILENAME SPI_REGION_OFFSETS SPI_REGION_SIZES
+	hl_check_length_of_spi_files  REGION_TO_FILENAME SPI_REGION_SIZES
+	hl_build_spi_flash_file REGION_TO_FILENAME SPI_REGION_OFFSETS SPI_REGION_SIZES OUTF
+}
+
diff --git a/board/habanalabs/jgs/spi-flash-layout-validator.sh b/board/habanalabs/jgs/spi-flash-layout-validator.sh
new file mode 100644
index 0000000..d7a13cb
--- /dev/null
+++ b/board/habanalabs/jgs/spi-flash-layout-validator.sh
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+
+function validate_spi_flash_layout {
+	local PPBOOT_BIN=${1}
+	local PREBOOT_BIN=${2}
+	local IMG_PTR_BIN=${3}
+	local FAKE_FUSE_BIN=${4}
+	declare -rA REGION_TO_FILENAME=(["IMAGE_DESC_PRIMARY"]=$IMG_PTR_BIN \
+					["PPBOOT_PRIMARY"]=$PPBOOT_BIN \
+					["PREBOOT_IO_DIE_PRIMARY"]=$PREBOOT_BIN \
+					["PREBOOT_COMPUTE_DIE_PRIMARY"]="FF" \
+					["CXL_ARC_ICCM_PRIMARY"]="FF" \
+					["CXL_ARC_DCCM_PRIMARY"]="FF" \
+					["DSM_ARC_ICCM_PRIMARY"]="FF" \
+					["DSM_ARC_DCCM_PRIMARY"]="FF" \
+					["RECOVERY_IMAGE_PCIE_PRIMARY"]="FF" \
+					["RECOVERY_IMAGE_FW_UPD_PRIMARY"]="FF" \
+					["FAKE_FUZE_PRIMARY"]=$FAKE_FUSE_BIN \
+					["RESERVED_1"]="FF" \
+					["IMAGE_DESC_SECONDARY"]=$IMG_PTR_BIN \
+					["PPBOOT_SECONDARY"]=$PPBOOT_BIN \
+					["PREBOOT_IO_DIE_SECONDARY"]=$PREBOOT_BIN \
+					["PREBOOT_COMPUTE_DIE_SECONDARY"]="FF" \
+					["CXL_ARC_ICCM_SECONDARY"]="FF" \
+					["CXL_ARC_DCCM_SECONDARY"]="FF" \
+					["DSM_ARC_ICCM_SECONDARY"]="FF" \
+					["DSM_ARC_DCCM_SECONDARY"]="FF" \
+					["RECOVERY_IMAGE_PCIE_SECONDARY"]="FF" \
+					["RECOVERY_IMAGE_FW_UPD_SECONDARY"]="FF" \
+					["FAKE_FUZE_SECONDARY"]=$FAKE_FUSE_BIN \
+					["RESERVED_2"]="FF")
+	declare -A SPI_REGION_OFFSETS
+	declare -A SPI_REGION_SIZES
+
+	hl_parse_spi_flash_regions SPI_REGION_OFFSETS SPI_REGION_SIZES fs1
+	hl_validate_spi_flash_layout REGION_TO_FILENAME SPI_REGION_OFFSETS SPI_REGION_SIZES
+	hl_check_length_of_spi_files  REGION_TO_FILENAME SPI_REGION_SIZES
+}
\ No newline at end of file
diff --git a/configs/jgs_asic_defconfig b/configs/jgs_asic_defconfig
new file mode 100644
index 0000000..4074539
--- /dev/null
+++ b/configs/jgs_asic_defconfig
@@ -0,0 +1,27 @@
+BR2_USE_MMU=y
+BR2_arcle=y
+BR2_archs4x=y
+BR2_SVN="svn"
+BR2_DL_DIR="$(BR2_EXTERNAL_HABANALABS_PATH)/dl"
+BR2_SHARED_STATIC_LIBS=y
+BR2_PACKAGE_OVERRIDE_FILE="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/jgs/jgs-local.mk"
+BR2_GLOBAL_PATCH_DIR="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/patches"
+BR2_TOOLCHAIN_EXTERNAL=y
+BR2_TARGET_GENERIC_HOSTNAME=""
+BR2_TARGET_GENERIC_ISSUE=""
+BR2_INIT_NONE=y
+BR2_SYSTEM_BIN_SH_NONE=y
+# BR2_TARGET_GENERIC_GETTY is not set
+BR2_PACKAGE_PREBOOT=y
+BR2_PIC_PIE=n
+BR2_SSP_NONE=y
+BR2_RELRO_NONE=y
+BR2_FORTIFY_SOURCE_NONE=y
+BR2_ROOTFS_POST_IMAGE_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/jgs/jgs-post-image-fit-arc.sh $(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/jgs/jgs-post-image-spi-arc_global.sh $(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/jgs/post-image-agents.sh"
+BR2_PACKAGE_BAREMETAL=y
+BR2_PACKAGE_HOST_UBOOT_TOOLS=y
+BR2_PACKAGE_HOST_UBOOT_TOOLS_FIT_SUPPORT=y
+BR2_PACKAGE_HOST_UBOOT_TOOLS_FIT_SIGNATURE_SUPPORT=y
+BR2_PACKAGE_ARCBOOT=y
+BR2_PACKAGE_MGMT=y
+BR2_PACKAGE_RESET=y
diff --git a/package/arcboot/arcboot.mk b/package/arcboot/arcboot.mk
index 64efc72..6b9627b 100644
--- a/package/arcboot/arcboot.mk
+++ b/package/arcboot/arcboot.mk
@@ -12,6 +12,10 @@
 ROM_BIN = boot_rom
 endif
 
+ifeq ($(HL_DEV),jgs)
+ROM_BIN = boot_rom
+endif
+
 ifeq ($(wildcard $(ARCBOOT_SRC_DIR)/bin/$(HL_DEV)/$(HL_TARGET)/$(ROM_BIN).bin),)
 define ARCBOOT_BUILD_CMDS
 	$(TARGET_MAKE_ENV) $(MAKE) CROSS_COMPILE="$(TARGET_CROSS)" -C $(@D) all
diff --git a/package/zephyr/jgs.mk b/package/zephyr/jgs.mk
new file mode 100644
index 0000000..3a97b78
--- /dev/null
+++ b/package/zephyr/jgs.mk
@@ -0,0 +1,18 @@
+#Buildroot is including all the makefiles for
+#the compilation of zephyr appliation Hence
+#adding additional check for fs1 to avoid
+#including SPECS_H17_DIR for other platforms
+
+ifeq ($(DEV_PREFIX),jgs)
+export DEVICE_SPECS_ROOT = $(SPECS_H17_DIR)
+export DEVICE_SPECS_DIR = $(DEVICE_SPECS_ROOT)/jgs
+export SPECS_EMB_PROJ_DIR = $(SPECS_EMBEDDED_ROOT)/fs/
+export ZEPHYR_CONF_DEV_EXTRA_CFLAGS =  -I$(DEVICE_SPECS_ROOT)
+export ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(DEVICE_SPECS_DIR)/asic_reg
+export ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(SPECS_EMB_PROJ_DIR)/common
+export ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(SPECS_EMB_PROJ_DIR)/$(DEV_PREFIX)
+export ZEPHYR_BOARD_PREFIX = $(DEVICE_NAME)
+
+/* TODO: remove this once pid application is ported */
+export ZEPHYR_PID_APP_DIR = $(EMBEDDED_DIR)/zephyr-apps/preboot/app/preboot_d1
+endif
diff --git a/scripts/make/hl-common/Makefile b/scripts/make/hl-common/Makefile
index b93e746..055c5df 100644
--- a/scripts/make/hl-common/Makefile
+++ b/scripts/make/hl-common/Makefile
@@ -1,6 +1,6 @@
 HL_MULTI_ARCH_RELEASE := gaudi2 gaudi3 fs1
 HL_DEFAULT_DEV_ARCH := gaudi2_arm gaudi3_arc fs1_arc
-HL_SINGLE_ARCH_RELEASE :=
+HL_SINGLE_ARCH_RELEASE := jgs
 HL_SECURITY_RELEASE :=
 
 RELEASE_JOB_FAKE_SEC := "fake_security=true"
@@ -149,7 +149,7 @@
 .PHONY: %-custom
 %-custom: ;
 
-HL_SEC_FIT_CREATE_ARCH := gaudi2 gaudi3 fs1
+HL_SEC_FIT_CREATE_ARCH := gaudi2 gaudi3 fs1 jgs
 
 .PHONY: %-key-revoke-create
 %-key-revoke-create:
diff --git a/scripts/make/jgs/Makefile b/scripts/make/jgs/Makefile
new file mode 100644
index 0000000..b3197cc
--- /dev/null
+++ b/scripts/make/jgs/Makefile
@@ -0,0 +1,93 @@
+export SPECS_H17_DIR ?= $(SPECS_ROOT_H17)
+
+JGS_RELEASE_REPOS_DIRS := embedded/buildroot embedded/buildroot-external \
+	embedded/fw_loader embedded/preboot embedded/robot \
+	embedded/hl-smi embedded/eeprom-util \
+	embedded/specs embedded/arcboot \
+	embedded/zephyr-apps/arcpid embedded/zephyr-apps/arcmgmt \
+	embedded/zephyr embedded/zephyr-modules/mbedtls embedded/zephyr-modules/fdt \
+	embedded/zephyr-modules/eeprom embedded/zephyr-modules/lz4 embedded/zephyr-plugins embedded/zephyr-modules/boot_info \
+	embedded/zephyr-apps/preboot embedded/zephyr-modules/soc_init \
+	embedded/zephyr-modules/hl_shell embedded/hl-test embedded/zephyr-modules/common \
+	embedded/zephyr-apps/pcie-ss embedded/zephyr-modules/spdm/hlspdm \
+	embedded/zephyr-modules/spdm/spdm-emu embedded/zephyr-modules/spdm/libspdm embedded/zephyr-apps/reset
+
+JGS_RELEASE_REPOS := buildroot buildroot-external fw_loader \
+	preboot robot hl-smi eeprom-util \
+	embedded-specs arcboot arcpid arcmgmt \
+	zephyr mbedtls fdt eeprom boot_info lz4 zephyr-plugins \
+	preboot_rtos soc_init hl_shell embedded-test emb-common-mods \
+	pcie-ss spdm spdm-emu libspdm reset
+
+JGS_INTERNAL_FITS := ${DEV_PREFIX}-security_enable.itb	\
+	${DEV_PREFIX}-secureboot_config.itb		\
+	${DEV_PREFIX}-secureboot_config_dbg.itb		\
+	${DEV_PREFIX}-secureboot_config_no_rev.itb	\
+	${DEV_PREFIX}-write_tpm_devid.itb		\
+	${DEV_PREFIX}-erom_recovery_enable.itb		\
+	${DEV_PREFIX}-erom_erase.itb			\
+	${DEV_PREFIX}-bis_compliant_enable.itb		\
+	${DEV_PREFIX}-slt_set.itb			\
+	${DEV_PREFIX}-rr_dump.itb
+
+.PHONY: jgs-internal-fit-create
+jgs-internal-fit-create: jgs-security-enable-create				\
+				jgs-secureboot-config-create			\
+				jgs-secureboot-config-dbg-create		\
+				jgs-secureboot-config-no-rev-create		\
+				jgs-write-tpm-devid-create			\
+				jgs-erom-recovery-enable-create			\
+				jgs-erom-erase-create				\
+				jgs-bis-compliant-enable-create			\
+				jgs-slt-set-create				\
+				jgs-rr-dump-create
+
+	@$(call hl-check-files,$(JGS_INTERNAL_FITS))
+	@tar czf $(IMAGES_DIR)/${DEV_PREFIX}-int-fit.tgz -C $(ITBS_DIR)/ $(JGS_INTERNAL_FITS)
+
+.PHONY: jgs-nic-sec-create
+jgs-nic-sec-create: hl-create-nic-default-ffffff
+
+JGS_SECURITY_FITS := ${DEV_PREFIX}-eeprom_read.itb	\
+	${DEV_PREFIX}-revoke_key.itb			\
+	${DEV_PREFIX}-security_status.itb		\
+	${DEV_PREFIX}-update_svn.itb			\
+	${DEV_PREFIX}-fw_upd_closure.itb		\
+	${DEV_PREFIX}-sram_sf-open.itb			\
+	${DEV_PREFIX}-sram_sf-close.itb			\
+	${DEV_PREFIX}-rma_erase.itb			\
+	${DEV_PREFIX}-rma_status.itb			\
+	${DEV_PREFIX}-efuse_read.itb
+
+.PHONY: jgs-security-fit-create
+jgs-security-fit-create: jgs-eeprom-read-create						\
+				jgs-key-revoke-create jgs-security-status-create	\
+				jgs-update-svn-create jgs-fw-upd-closure-create		\
+				jgs-sram-sf-access-create				\
+				jgs-rma-erase-create jgs-rma-status-create		\
+				jgs-efuse-read-create
+
+	@$(call hl-check-files,$(JGS_SECURITY_FITS))
+	@tar czf $(IMAGES_DIR)/${DEV_PREFIX}-sec-fit.tgz -C $(ITBS_DIR)/ $(JGS_SECURITY_FITS)
+
+.PHONY: jgs-print-release-repos
+jgs-print-release-repos:
+	@echo $(JGS_RELEASE_REPOS)
+
+.PHONY: jgs-print-release-dirs
+jgs-print-release-dirs:
+	@echo $(JGS_RELEASE_REPOS_DIRS)
+
+.PHONY: jgs-post-image
+jgs-post-image:
+	@echo "Creating jgs-spi-flash.bin"
+	@$(MAKE) jgs-bin-create
+
+ifeq ($(HL_TARGET),palad)
+	@echo "Convert SPI flash bin image into Palladium hex format"
+	source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/jgs-image-utility.sh && \
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.be && \
+	hl_palad_flash_hex spi-flash ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.be ${HEXS_DIR}
+endif
+
+JGS_BIN_FITS = ${DEV_PREFIX}-agent-fw_loader-fit.itb
